
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>adobo package &#8212; adobo  documentation</title>
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">adobo package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-adobo.IO">adobo.IO module</a><ul>
<li><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.bio">adobo.bio module</a><ul>
<li><a class="reference internal" href="#id1">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.clustering">adobo.clustering module</a><ul>
<li><a class="reference internal" href="#id2">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.data">adobo.data module</a><ul>
<li><a class="reference internal" href="#id5">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.dr">adobo.dr module</a><ul>
<li><a class="reference internal" href="#id6">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.hvg">adobo.hvg module</a><ul>
<li><a class="reference internal" href="#id7">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.normalize">adobo.normalize module</a><ul>
<li><a class="reference internal" href="#id8">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.plotting">adobo.plotting module</a><ul>
<li><a class="reference internal" href="#id9">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.preproc">adobo.preproc module</a><ul>
<li><a class="reference internal" href="#id10">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/adobo.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="adobo-package">
<h1>adobo package<a class="headerlink" href="#adobo-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="adobo.glm.html">adobo.glm package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.families">adobo.glm.families module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.glm">adobo.glm.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.glmnet">adobo.glm.glmnet module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.simulation">adobo.glm.simulation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.utils">adobo.glm.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="adobo.irlbpy.html">adobo.irlbpy package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#module-adobo.irlbpy.irlb">adobo.irlbpy.irlb module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#module-adobo.irlbpy">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-adobo.IO">
<span id="adobo-io-module"></span><h2>adobo.IO module<a class="headerlink" href="#module-adobo.IO" title="Permalink to this headline">¶</a></h2>
<div class="section" id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<p>Functions for reading and writing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.IO.load_from_file">
<code class="sig-prename descclassname">adobo.IO.</code><code class="sig-name descname">load_from_file</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">sep='\t'</em>, <em class="sig-param">header=0</em>, <em class="sig-param">column_id='auto'</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">desc='no desc set'</em>, <em class="sig-param">save_filename=None</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.IO.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a gene expression matrix consisting of raw read counts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<cite>str</cite>) – Path to the file containing input data. Should be a matrix where
columns are cells and rows are genes.</p></li>
<li><p><strong>sep</strong> (<cite>str</cite>) – Character used to separate fields. Default: “\t”</p></li>
<li><p><strong>header</strong> (<cite>str</cite>) – If the data file has a header. 0 means yes otherwise None. Default: 0</p></li>
<li><p><strong>column_id</strong> (<em>{'auto'</em><em>, </em><em>'yes'</em><em>, </em><em>'no'}</em>) – Whether the header (first line) of the file contains a column ID for the genes. If
this is the case, set this to auto or yes, otherwise no. Default: ‘auto’</p></li>
<li><p><strong>desc</strong> (<cite>str</cite>) – A description of the data</p></li>
<li><p><strong>save_filename</strong> (<cite>str</cite>) – An output filename used when calling <a class="reference internal" href="#adobo.data.dataset.save" title="adobo.data.dataset.save"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.data.dataset.save()</span></code></a>.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – To be verbose or not. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The loaded gene expression matrix should not have been normalized. This function calls
<code class="xref py py-func docutils literal notranslate"><span class="pre">read_csv()</span></code> to read the data matrix file. Any additional
arguments are passed into <code class="xref py py-func docutils literal notranslate"><span class="pre">read_csv()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dataset class object.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.bio">
<span id="adobo-bio-module"></span><h2>adobo.bio module<a class="headerlink" href="#module-adobo.bio" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Summary<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Functions related to biology.</p>
<dl class="function">
<dt id="adobo.bio.cell_cycle_predict">
<code class="sig-prename descclassname">adobo.bio.</code><code class="sig-name descname">cell_cycle_predict</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">genes_s_phase</em>, <em class="sig-param">genes_g2m_phase</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.bio.cell_cycle_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts cell cycle phase</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>genes_s_phase</strong> (<cite>list</cite>) – A list of S phase genes.</p></li>
<li><p><strong>genes_g2m_phase</strong> (<cite>list</cite>) – A list of G2/M phase genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.clustering">
<span id="adobo-clustering-module"></span><h2>adobo.clustering module<a class="headerlink" href="#module-adobo.clustering" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id2">
<h3>Summary<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>This module contains functions to cluster data.</p>
<dl class="function">
<dt id="adobo.clustering.generate">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">generate</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">k=10</em>, <em class="sig-param">graph='snn'</em>, <em class="sig-param">clust_alg='leiden'</em>, <em class="sig-param">prune_snn=0.067</em>, <em class="sig-param">res=0.8</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.generate" title="Permalink to this definition">¶</a></dt>
<dd><p>A wrapper function for generating single cell clusters from a shared nearest neighbor
graph with the Leiden algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>k</strong> (<cite>int</cite>) – Number of nearest neighbors. Default: 10</p></li>
<li><p><strong>graph</strong> (<cite>{‘snn’}</cite>) – Type of graph to generate. Only shared nearest neighbor (snn) supported at the
moment.</p></li>
<li><p><strong>clust_alg</strong> (<a href="#id3"><span class="problematic" id="id4">`</span></a>{‘leiden’, ‘walktrap’, ‘spinglass’, ‘multilevel’, ‘infomap’,) – ‘label_prop’, ‘leading_eigenvector’}`
Clustering algorithm to be used.</p></li>
<li><p><strong>prune_snn</strong> (<cite>float</cite>) – Threshold for pruning the SNN graph, i.e. the edges  with lower value (Jaccard
index) than this will be removed. Set to 0 to disable pruning. Increasing this
value will result in fewer edges in the graph. Default: 0.067</p></li>
<li><p><strong>res</strong> (<cite>float</cite>) – Resolution parameter for the Leiden algorithm _only_; change to modify cluster
resolution. Default: 0.8</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="simple">
<dt>Yang et al. (2016) A Comparative Analysis of Community Detection Algorithms on</dt><dd><p>Artificial Networks. Scientific Reports</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.data">
<span id="adobo-data-module"></span><h2>adobo.data module<a class="headerlink" href="#module-adobo.data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id5">
<h3>Summary<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>This module contains a data storage class.</p>
<dl class="class">
<dt id="adobo.data.dataset">
<em class="property">class </em><code class="sig-prename descclassname">adobo.data.</code><code class="sig-name descname">dataset</code><span class="sig-paren">(</span><em class="sig-param">raw_mat</em>, <em class="sig-param">desc='no desc set'</em>, <em class="sig-param">save_filename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Storage container for raw data and analysis results.</p>
<dl class="attribute">
<dt id="adobo.data.dataset._assays">
<code class="sig-name descname">_assays</code><a class="headerlink" href="#adobo.data.dataset._assays" title="Permalink to this definition">¶</a></dt>
<dd><p>Holding information about what functions have been applied.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.exp_mat">
<code class="sig-name descname">exp_mat</code><a class="headerlink" href="#adobo.data.dataset.exp_mat" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw read count matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset._low_quality_cells">
<code class="sig-name descname">_low_quality_cells</code><a class="headerlink" href="#adobo.data.dataset._low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Low quality cells identified with <a class="reference internal" href="#adobo.preproc.find_low_quality_cells" title="adobo.preproc.find_low_quality_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adobo.preproc.find_low_quality_cells()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.norm_log2">
<code class="sig-name descname">norm_log2</code><a class="headerlink" href="#adobo.data.dataset.norm_log2" title="Permalink to this definition">¶</a></dt>
<dd><p>True if log2 transform was performed on the normalized data, otherwise False.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.norm_method">
<code class="sig-name descname">norm_method</code><a class="headerlink" href="#adobo.data.dataset.norm_method" title="Permalink to this definition">¶</a></dt>
<dd><p>Containing the method used for normalization.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.norm">
<code class="sig-name descname">norm</code><a class="headerlink" href="#adobo.data.dataset.norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalized gene expression data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.hvg">
<code class="sig-name descname">hvg</code><a class="headerlink" href="#adobo.data.dataset.hvg" title="Permalink to this definition">¶</a></dt>
<dd><p>Containing highly variable genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.dr">
<code class="sig-name descname">dr</code><a class="headerlink" href="#adobo.data.dataset.dr" title="Permalink to this definition">¶</a></dt>
<dd><p>A dict of <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing components from dimensionality
reduction.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.dr_gene_contr">
<code class="sig-name descname">dr_gene_contr</code><a class="headerlink" href="#adobo.data.dataset.dr_gene_contr" title="Permalink to this definition">¶</a></dt>
<dd><p>A dict of <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> containing variable contributions to each
component. Useful for understanding the contribution of each gene to PCA
components.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.meta_cells">
<code class="sig-name descname">meta_cells</code><a class="headerlink" href="#adobo.data.dataset.meta_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>A data frame containing meta data for cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>pandas.DataFrame</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.meta_genes">
<code class="sig-name descname">meta_genes</code><a class="headerlink" href="#adobo.data.dataset.meta_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>A data frame containing meta data for genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>pandas.DataFrame</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.clusters">
<code class="sig-name descname">clusters</code><a class="headerlink" href="#adobo.data.dataset.clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Generated clusters.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.desc">
<code class="sig-name descname">desc</code><a class="headerlink" href="#adobo.data.dataset.desc" title="Permalink to this definition">¶</a></dt>
<dd><p>A string describing the dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.add_meta_data">
<code class="sig-name descname">add_meta_data</code><span class="sig-paren">(</span><em class="sig-param">axis</em>, <em class="sig-param">key</em>, <em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.add_meta_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Add meta data to the adobo object</p>
<p class="rubric">Notes</p>
<p>Meta data can be added to cells or genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<cite>{‘cells’, ‘genes’}</cite>) – Should the data be added to cells or genes?</p></li>
<li><p><strong>key</strong> (<cite>str</cite>) – The variable name for your data. No whitespaces and special characters.</p></li>
<li><p><strong>data</strong> (<cite>list</cite>) – A list of data to add. The length must match the length of cells or genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.assays">
<code class="sig-name descname">assays</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.assays" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays a basic summary of the dataset and what analyses have been performed on
it.</p>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.get_assay">
<code class="sig-name descname">get_assay</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">lang=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.get_assay" title="Permalink to this definition">¶</a></dt>
<dd><p>Get info if a function has been applied.</p>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.low_quality_cells">
<em class="property">property </em><code class="sig-name descname">low_quality_cells</code><a class="headerlink" href="#adobo.data.dataset.low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">compress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the object</p>
<p class="rubric">Notes</p>
<p>Load the object data with joblib.load</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>compress</strong> (<cite>bool</cite>) – Save with data compression or not. Default: True</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.set_assay">
<code class="sig-name descname">set_assay</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">key=1</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.set_assay" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the assay that was applied.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.dr">
<span id="adobo-dr-module"></span><h2>adobo.dr module<a class="headerlink" href="#module-adobo.dr" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id6">
<h3>Summary<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<p>Functions for dimensional reduction.</p>
<dl class="function">
<dt id="adobo.dr.irlb">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">irlb</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ncomp=75</em>, <em class="sig-param">seed=42</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.irlb" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncated SVD by implicitly restarted Lanczos bidiagonalization</p>
<p class="rubric">Notes</p>
<p>The augmented implicitly restarted Lanczos bidiagonalization algorithm (IRLBA) finds
a few approximate largest singular values and corresponding singular vectors using a
method of Baglama and Reichel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return, optional.</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility, optional.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Baglama et al (2005) Augmented Implicitly Restarted Lanczos Bidiagonalization Methods
SIAM Journal on Scientific Computing</p>
<p><a class="reference external" href="https://github.com/bwlewis/irlbpy">https://github.com/bwlewis/irlbpy</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the components (columns).</p></li>
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the contributions of every gene (rows).</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.pca">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">pca</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='irlb'</em>, <em class="sig-param">ncomp=75</em>, <em class="sig-param">allgenes=False</em>, <em class="sig-param">scale=True</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">seed=42</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Principal Component Analysis</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual normalization functions, which can also be
called directly. Scaling of the data is achieved by setting scale=True (default),
which will center (subtract the column mean) and scale columns (divide by their
standard deviation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘irlb’, ‘svd’}</cite>) – Method to use for PCA. This does not matter much. Default: irlb</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return. Default: 75</p></li>
<li><p><strong>allgenes</strong> (<cite>bool</cite>) – Use all genes instead of only HVG. Default: False</p></li>
<li><p><strong>scale</strong> (<cite>bool</cite>) – Scales input data prior to PCA. Default: True</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be noisy or not. Default: False</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility (only irlb). Default: 42</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Principal_component_analysis">https://en.wikipedia.org/wiki/Principal_component_analysis</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>Nothing. Modifies the passed object. Results are stored in two dictonaries in the</em></p></li>
<li><p><strong>passed object</strong> (<cite>dr</cite> (containing the components) and <cite>dr_gene_contr</cite> (containing)</p></li>
<li><p><em>gene contributions to each component)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.svd">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">svd</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ncomp=75</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.svd" title="Permalink to this definition">¶</a></dt>
<dd><p>Principal component analysis via singular value decomposition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data. Preferrably this
should be a subset of the normalized gene expression matrix containing highly
variable genes.</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return, optional.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>(SE) <a class="reference external" href="https://tinyurl.com/yyt6df5x">https://tinyurl.com/yyt6df5x</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the components (columns).</p></li>
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the contributions of every gene (rows).</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.tsne">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">tsne</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">target='irlb'</em>, <em class="sig-param">perplexity=30</em>, <em class="sig-param">n_iter=2000</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Projects data to a two dimensional space using the tSNE algorithm.</p>
<p class="rubric">Notes</p>
<p>It is recommended to perform this function on data in PCA space. This function calls
<code class="xref py py-func docutils literal notranslate"><span class="pre">sklearn.manifold.TSNE()</span></code>, and any additional parameters will be passed to it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>target</strong> (<cite>{‘irlb’, ‘svd’, ‘norm’}</cite>) – What to run tSNE on. Default: ‘irlb’</p></li>
<li><p><strong>perplexity</strong> (<cite>float</cite>) – From [1]: The perplexity is related to the number of nearest neighbors that
is used in other manifold learning algorithms. Larger datasets usually require
a larger perplexity. Consider selecting a value between 5 and 50. Different
values can result in significanlty different results. Default: 30</p></li>
<li><p><strong>n_iter</strong> (<cite>int</cite>) – Number of iterations. Default: 2000</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility. Default: 42</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>[0] van der Maaten, L.J.P.; Hinton, G.E. Visualizing High-Dimensional Data
Using t-SNE. Journal of Machine Learning Research 9:2579-2605, 2008.</p>
<p>[1] <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html">https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.umap">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">umap</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">target='irlb'</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Projects data to a two dimensional space using the UMAP algorithm, a non-linear
data reduction algorithm.</p>
<p class="rubric">Notes</p>
<p>The UMAP output can be tweaked using these parameters (but there are also several
other parameters that can influence the outcome, see <cite>help(umap.UMAP)</cite>:
<cite>n_neighbors</cite>, Default: 15</p>
<blockquote>
<div><p>This parameter controls the balances between local versus global structure.</p>
</div></blockquote>
<dl class="simple">
<dt><cite>min_dist</cite>, Default: 0.1</dt><dd><p>Controls how tightly points are packed together.</p>
</dd>
<dt><cite>metric</cite>, Default: ‘euclidean’</dt><dd><p>The metric to use to compute distances in high dimensional space.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>target</strong> (<cite>{‘irlb’, ‘svd’, ‘norm’}</cite>) – What to run tSNE on.</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>McInnes L, Healy J, Melville J, arxiv, 2018</p>
<p><a class="reference external" href="https://arxiv.org/abs/1802.03426">https://arxiv.org/abs/1802.03426</a>
<a class="reference external" href="https://github.com/lmcinnes/umap">https://github.com/lmcinnes/umap</a>
<a class="reference external" href="https://umap-learn.readthedocs.io/en/latest/">https://umap-learn.readthedocs.io/en/latest/</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.hvg">
<span id="adobo-hvg-module"></span><h2>adobo.hvg module<a class="headerlink" href="#module-adobo.hvg" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id7">
<h3>Summary<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<p>Functions for detection of highly variable genes.</p>
<dl class="function">
<dt id="adobo.hvg.brennecke">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">brennecke</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log2</em>, <em class="sig-param">ercc=Empty DataFrame Columns: [] Index: []</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">minBiolDisp=0.5</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.brennecke" title="Permalink to this definition">¶</a></dt>
<dd><p>Implements the method of Brennecke et al. (2013) to identify highly variable genes</p>
<p class="rubric">Notes</p>
<p>Fits data using GLM with Fisher Scoring. GLM code copied from (credits to &#64;madrury
for this code): <a class="reference external" href="https://github.com/madrury/py-glm">https://github.com/madrury/py-glm</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>log2</strong> (<cite>bool</cite>) – If normalized data were log2 transformed or not.</p></li>
<li><p><strong>ercc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized ercc spikes.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>minBiolDisp</strong> (<cite>float</cite>) – Minimum percentage of variance due to biological factors.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Brennecke et al. (2013) Nature Methods <a class="reference external" href="https://doi.org/10.1038/nmeth.2645">https://doi.org/10.1038/nmeth.2645</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.chen2016">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">chen2016</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log2</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.chen2016" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from Chen (2016) to identify highly variable
genes.</p>
<p class="rubric">Notes</p>
<p>Expression counts should be normalized and on a log scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>log2</strong> (<cite>bool</cite>) – If normalized data were log2 transformed or not.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-2897-6">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-2897-6</a>
<a class="reference external" href="https://github.com/hillas/scVEGs/blob/master/scVEGs.r">https://github.com/hillas/scVEGs/blob/master/scVEGs.r</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.find_hvg">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">find_hvg</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='seurat'</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.find_hvg" title="Permalink to this definition">¶</a></dt>
<dd><p>Finding highly variable genes</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual HVG functions, which can also be called
directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘seurat’, ‘brennecke’, ‘scran’, ‘chen2016’, ‘mm’}</cite>) – Specifies the method to be used.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of genes to return.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate threshold for significant genes applied to those methods
that use it (brennecke, chen2016, mm). Note that the number of returned genes
might be fewer than specified by <cite>ngenes</cite> because of FDR consideration.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Yip et al. (2018) Briefings in Bioinformatics
<a class="reference external" href="https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby011/4898116">https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby011/4898116</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.mm">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">mm</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log2</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.mm" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from Andrews (2018).</p>
<p class="rubric">Notes</p>
<p>Input should be normalized but nog log’ed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing raw read counts.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty1044">https://doi.org/10.1093/bioinformatics/bty1044</a>
<a class="reference external" href="https://github.com/tallulandrews/M3Drop">https://github.com/tallulandrews/M3Drop</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.scran">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">scran</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ercc</em>, <em class="sig-param">log2</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.scran" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from the scran R package</p>
<p class="rubric">Notes</p>
<p>Expression counts should be normalized and on a log scale.</p>
<p>Outline of the steps:</p>
<ol class="arabic simple">
<li><p>fits a polynomial regression model to mean and variance of the technical genes</p></li>
<li><p>decomposes the total variance of the biological genes by subtracting the
technical variance predicted by the fit</p></li>
<li><p>sort based on biological variance</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>ercc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized ercc spikes.</p></li>
<li><p><strong>log2</strong> (<cite>bool</cite>) – If normalized data were log2 transformed or not.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Lun ATL, McCarthy DJ, Marioni JC (2016). “A step-by-step workflow for low-level
analysis of single-cell RNA-seq data with Bioconductor.” F1000Research,
doi.org/10.12688/f1000research.9501.2</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.seurat">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">seurat</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">num_bins=20</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.seurat" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves a list of highly variable genes using Seurat’s strategy</p>
<p class="rubric">Notes</p>
<p>The function bins the genes according to average expression, then calculates
dispersion for each bin as variance to mean ratio. Within each bin, Z-scores are
calculated and returned. Z-scores are ranked and the top 1000 are selected. Input data
should be normalized first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
<li><p><strong>num_bins</strong> (<cite>int</cite>) – Number of bins to use.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p><a class="reference external" href="https://cran.r-project.org/web/packages/Seurat/index.html">https://cran.r-project.org/web/packages/Seurat/index.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.normalize">
<span id="adobo-normalize-module"></span><h2>adobo.normalize module<a class="headerlink" href="#module-adobo.normalize" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id8">
<h3>Summary<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>This module contains functions to normalize raw read counts.</p>
<dl class="function">
<dt id="adobo.normalize.clr">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">clr</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">axis='genes'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.clr" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs centered log ratio normalization similar to Seurat</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>axis</strong> (<em>{'genes'</em><em>, </em><em>'cells'}</em>) – Normalize over genes or cells. Default: ‘genes’</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Hafemeister et al. (2019) <a class="reference external" href="https://www.biorxiv.org/content/10.1101/576827v1">https://www.biorxiv.org/content/10.1101/576827v1</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.fqn">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">fqn</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.fqn" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs full quantile normalization (FQN)</p>
<p class="rubric">Notes</p>
<p>FQN was shown to perform well on single cell data [1] and was a popular
normalization scheme for microarray data. The present function does not handle ties
well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="simple">
<dt>[0] Bolstad et al. (2003) Bioinformatics</dt><dd><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/19/2/185/372664">https://academic.oup.com/bioinformatics/article/19/2/185/372664</a></p>
</dd>
<dt>[1] Cole et al. (2019) Cell Systems</dt><dd><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/235382v2">https://www.biorxiv.org/content/10.1101/235382v2</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.norm">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">norm</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='standard'</em>, <em class="sig-param">log2=True</em>, <em class="sig-param">small_const=1</em>, <em class="sig-param">remove_low_qual=True</em>, <em class="sig-param">gene_lengths=None</em>, <em class="sig-param">scaling_factor=10000</em>, <em class="sig-param">axis='genes'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalizes gene expression data</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual normalization functions, which can also be
called directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘standard’, ‘rpkm’, ‘fqn’, ‘clr’, ‘vsn’}</cite>) – Specifies the method to use. <cite>standard</cite> refers to the simplest normalization
strategy involving scaling genes by total number of reads per cell. <cite>rpkm</cite>
performs RPKM normalization and requires the <cite>gene_lengths</cite> parameter to be set.
<cite>fqn</cite> performs a full-quantile normalization. <cite>clr</cite> performs centered log ratio
normalization. <cite>vsn</cite> performs a variance stabilizing normalization.
Default: standard</p></li>
<li><p><strong>log2</strong> (<cite>bool</cite>) – Perform log2 transformation. Not applicable if method=’vsn’. Default: True</p></li>
<li><p><strong>small_const</strong> (<cite>float</cite>) – A small constant to add to expression values to avoid log’ing genes with zero
expression. Default: 1</p></li>
<li><p><strong>remove_low_qual</strong> (<cite>bool</cite>) – Remove low quality cells and uninformative genes identified by prior steps.
Default: True</p></li>
<li><p><strong>gene_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> or <cite>str</cite>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> containing the gene lengths in base pairs and gene names
set as index. The names must match the gene names used in <cite>data</cite> (the order does
not need to match and any symbols not found in the data will be discarded).
Normally gene lengths should be the combined length of exons for every gene. If
gene_lengths is a <cite>str</cite> then it is taken as a filename and loaded; first column is
gene names and second column is the length, field separator is one space.
<cite>gene_lengths</cite> needs to be set _only_ if method=’rpkm’. Default: None</p></li>
<li><p><strong>scaling_factor</strong> (<cite>int</cite>) – Scaling factor used to multiply the scaled counts with. Only used for
<cite>method=”depth”</cite>. Default: 10000</p></li>
<li><p><strong>axis</strong> (<em>{'genes'</em><em>, </em><em>'cells'}</em>) – Only applicable when <cite>method=”clr”</cite>, defines the axis to normalize across.
Default: ‘genes’</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="simple">
<dt>[0] Cole et al. (2019) Cell Systems</dt><dd><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/235382v2">https://www.biorxiv.org/content/10.1101/235382v2</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#adobo.normalize.standard" title="adobo.normalize.standard"><code class="xref py py-func docutils literal notranslate"><span class="pre">standard()</span></code></a>, <a class="reference internal" href="#adobo.normalize.rpkm" title="adobo.normalize.rpkm"><code class="xref py py-func docutils literal notranslate"><span class="pre">rpkm()</span></code></a>, <a class="reference internal" href="#adobo.normalize.fqn" title="adobo.normalize.fqn"><code class="xref py py-func docutils literal notranslate"><span class="pre">fqn()</span></code></a>, <a class="reference internal" href="#adobo.normalize.clr" title="adobo.normalize.clr"><code class="xref py py-func docutils literal notranslate"><span class="pre">clr()</span></code></a>, <a class="reference internal" href="#adobo.normalize.vsn" title="adobo.normalize.vsn"><code class="xref py py-func docutils literal notranslate"><span class="pre">vsn()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.rpkm">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">rpkm</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">gene_lengths</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.rpkm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize expression values as RPKM</p>
<p class="rubric">Notes</p>
<p>This method should be used if you need to adjust for gene length, such as in
a SMART-Seq2 protocol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>gene_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> or <cite>str</cite>) – Should contain the gene lengths in base pairs and gene names set as index. The
names must match the gene names used in <cite>data</cite>. Normally gene lengths should be
the combined length of exons for every gene. If gene_lengths is a <cite>str</cite> then it is
taken as a filename and loaded; first column is gene names and second column is
the length, field separator is one space.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Conesa et al. (2016) Genome Biology
<a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.standard">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">standard</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">scaling_factor=10000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.standard" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs a standard normalization by scaling with the total read depth per cell and
then multiplying with a scaling factor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>scaling_factor</strong> (<cite>int</cite>) – Scaling factor used to multiply the scaled counts with. Default: 10000</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Evans et al. (2018) Briefings in Bioinformatics
<a class="reference external" href="https://academic.oup.com/bib/article/19/5/776/3056951">https://academic.oup.com/bib/article/19/5/776/3056951</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.vsn">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">vsn</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">min_cells=10</em>, <em class="sig-param">gmean_eps=1</em>, <em class="sig-param">n_genes=2000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.vsn" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs variance stabilizing normaliztion based on a negative binomial regression
model with regularized parameters</p>
<p class="rubric">Notes</p>
<p>Adopts a subset of the functionality of <cite>vst</cite> in the R package sctransform [0].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>min_cells</strong> (<cite>int</cite>) – Minimum number of cells expressing a gene for the gene to be used. Default: 10</p></li>
<li><p><strong>gmean_eps</strong> (<cite>float</cite>) – A small constant to avoid log(0)=-Inf. Default: 1</p></li>
<li><p><strong>n_genes</strong> (<cite>int</cite>) – Number of genes to use when estimating parameters. Default: 2000</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>[0] <a class="reference external" href="https://cran.r-project.org/web/packages/sctransform/index.html">https://cran.r-project.org/web/packages/sctransform/index.html</a>
[1] <a class="reference external" href="https://www.biorxiv.org/content/10.1101/576827v1">https://www.biorxiv.org/content/10.1101/576827v1</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.plotting">
<span id="adobo-plotting-module"></span><h2>adobo.plotting module<a class="headerlink" href="#module-adobo.plotting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id9">
<h3>Summary<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<p>Functions for plotting scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.plotting.cell_viz">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">cell_viz</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">reduction='tsne'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">marker_size=0.8</em>, <em class="sig-param">font_size=8</em>, <em class="sig-param">cluster_colors='adobo'</em>, <em class="sig-param">title=''</em>, <em class="sig-param">legend=True</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.cell_viz" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a 2d scatter plot from an embedding</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>reduction</strong> (<cite>{‘tsne’, ‘umap’, ‘irlb’, ‘svd’}</cite>) – The dimensional reduction to use. Default: tsne</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Name of an output file instead of showing on screen.</p></li>
<li><p><strong>marker_size</strong> (<cite>float</cite>) – The size of the markers.</p></li>
<li><p><strong>font_size</strong> (<cite>float</cite>) – Font size. Default: 8</p></li>
<li><p><strong>cluster_colors</strong> (<cite>{‘default’, ‘random’}</cite> or <cite>list</cite>) – Can be: (i) “adobo” or “random”; (ii) a list of colors with the same
length as the number of cells (same order as cells occur in the normalized
matrix). If cluster_colors is set to “adobo”, then colors are retrieved from
<code class="xref py py-attr docutils literal notranslate"><span class="pre">adobo._constants.CLUSTER_COLORS_DEFAULT</span></code> (but if the number of clusters
exceed 50, then random colors will be used). Default: adobo</p></li>
<li><p><strong>title</strong> (<cite>str</cite>) – Title of the plot.</p></li>
<li><p><strong>legend</strong> (<cite>bool</cite>) – Add legend or not. Default: True</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.genes_per_cell">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">genes_per_cell</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">barcolor='#E69F00'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">title='expressed genes'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.genes_per_cell" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a bar plot of number of expressed genes per cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>barcolor</strong> (<cite>str</cite>, optional) – Color of the bars (default: “#E69F00”).</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file.</p></li>
<li><p><strong>title</strong> (<cite>str</cite>, optional) – Title of the plot (default: “sequencing reads”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.pca_contributors">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">pca_contributors</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">target='irlb'</em>, <em class="sig-param">dim=range(0</em>, <em class="sig-param">5)</em>, <em class="sig-param">top=10</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">fontsize=8</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.pca_contributors" title="Permalink to this definition">¶</a></dt>
<dd><p>Examine the top contributing genes for each PCA component</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Genes are ranked by their absolute value. Additional parameters are passed into
<code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.savefig()</span></code>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>target</strong> (<cite>{‘irlb’, ‘svd’}</cite>) – The dimensional reduction to use. Default: irlb</p></li>
<li><p><strong>dim</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">range</span></code>) – Specifies the components to plot. For example: range(0,5) specifies the first five.</p></li>
<li><p><strong>top</strong> (<cite>int</cite>) – Specifies the number of top scoring genes to include. Default: 10</p></li>
<li><p><strong>fontsize</strong> (<cite>int</cite>) – Specifies font size. Default: 8</p></li>
<li><p><strong>color</strong> (<cite>str</cite>) – Color of the bars. As a string or hex code. Default: “#fcc603”</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write to a file instead of showing the plot on screen.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.reads_per_cell">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">reads_per_cell</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">barcolor='#E69F00'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">title='sequencing reads'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.reads_per_cell" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a bar plot of read counts per cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>barcolor</strong> (<cite>str</cite>, optional) – Color of the bars (default: “#E69F00”).</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file.</p></li>
<li><p><strong>title</strong> (<cite>str</cite>, optional) – Title of the plot (default: “sequencing reads”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.preproc">
<span id="adobo-preproc-module"></span><h2>adobo.preproc module<a class="headerlink" href="#module-adobo.preproc" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id10">
<h3>Summary<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<p>Functions for pre-processing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.preproc.detect_ercc_spikes">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">detect_ercc_spikes</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">ercc_pattern='^ERCC[_-]\\S+$'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.detect_ercc_spikes" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag ERCC spikes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>ercc_pattern</strong> (<cite>str</cite>, optional) – A regular expression matching ercc gene symbols. Default: “ercc[_-]S+$”</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of detected ercc spikes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.find_low_quality_cells">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_low_quality_cells</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">rRNA_genes</em>, <em class="sig-param">sd_thres=3</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Statistical detection of low quality cells using Mahalanobis distances</p>
<p class="rubric">Notes</p>
<p>Mahalanobis distances are computed from five quality metrics. A robust estimate of
covariance is used in the Mahalanobis function. Cells with Mahalanobis distances of
three standard deviations from the mean are by default considered outliers.
The five metrics are:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>log-transformed number of molecules detected</p></li>
<li><p>the number of genes detected</p></li>
<li><p>the percentage of reads mapping to ribosomal</p></li>
<li><p>mitochondrial genes</p></li>
<li><p>ercc recovery (if available)</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>rRNA_genes</strong> (<cite>list</cite> or <cite>str</cite>) – Either a list of rRNA genes or a string containing the path to a file containing
the rRNA genes (one gene per line).</p></li>
<li><p><strong>sd_thres</strong> (<cite>float</cite>) – Number of standard deviations to consider significant, i.e. cells are low quality
if this. Set to higher to remove fewer cells. Default: 3</p></li>
<li><p><strong>seed</strong> (<cite>float</cite>) – For the random number generator. Default: 42</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of low quality cells that were identified, and also modifies the passed
object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.find_mitochondrial_genes">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_mitochondrial_genes</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">mito_pattern='^mt-'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_mitochondrial_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Find mitochondrial genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>mito_pattern</strong> (<cite>str</cite>) – A regular expression matching mitochondrial gene symbols. Default: “^mt-“</p></li>
<li><p><strong>verbose</strong> (<em>boolean</em>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of mitochondrial genes detected.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.reset_filters">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">reset_filters</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.reset_filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets cell and gene filters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.set_mitochondrial_genes">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">set_mitochondrial_genes</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">genes</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.set_mitochondrial_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Set mitochondrial genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>genes</strong> (<cite>list</cite>) – A list of mitochondrial genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of genes found.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.simple_filter">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">simple_filter</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">minreads=1000</em>, <em class="sig-param">maxreads=None</em>, <em class="sig-param">minexpgenes=0.001</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.simple_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cells with too few reads and genes with very low expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>minreads</strong> (<cite>int</cite>, optional) – Minimum number of reads per cell required to keep the cell. Default: 1000</p></li>
<li><p><strong>maxreads</strong> (<cite>int</cite>, optional) – Set a maximum number of reads allowed. Useful for filtering out suspected doublets.</p></li>
<li><p><strong>minexpgenes</strong> (<cite>str</cite>, optional) – If this value is a float, then at least that fraction of cells must express the
gene. If integer, then it denotes the minimum that number of cells must express
the gene. Default: 0.001</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>int</em> – Number of cells removed.</p></li>
<li><p><em>int</em> – Number of genes removed.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-adobo" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="adobo.excepthook">
<code class="sig-prename descclassname">adobo.</code><code class="sig-name descname">excepthook</code><span class="sig-paren">(</span><em class="sig-param">type</em>, <em class="sig-param">value</em>, <em class="sig-param">tb</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.excepthook" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Oscar Franzen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>