
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>adobo package &#8212; adobo  documentation</title>
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">adobo package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-adobo.IO">adobo.IO module</a><ul>
<li><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.bio">adobo.bio module</a><ul>
<li><a class="reference internal" href="#id1">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.clustering">adobo.clustering module</a><ul>
<li><a class="reference internal" href="#id3">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.data">adobo.data module</a><ul>
<li><a class="reference internal" href="#id14">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.de">adobo.de module</a><ul>
<li><a class="reference internal" href="#id15">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.dr">adobo.dr module</a><ul>
<li><a class="reference internal" href="#id19">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.hvg">adobo.hvg module</a><ul>
<li><a class="reference internal" href="#id30">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.normalize">adobo.normalize module</a><ul>
<li><a class="reference internal" href="#id39">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.plotting">adobo.plotting module</a><ul>
<li><a class="reference internal" href="#id48">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.preproc">adobo.preproc module</a><ul>
<li><a class="reference internal" href="#id49">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.traj">adobo.traj module</a><ul>
<li><a class="reference internal" href="#id52">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/adobo.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="adobo-package">
<h1>adobo package<a class="headerlink" href="#adobo-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="adobo.glm.html">adobo.glm package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.families">adobo.glm.families module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.glm">adobo.glm.glm module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.glmnet">adobo.glm.glmnet module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.simulation">adobo.glm.simulation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm.utils">adobo.glm.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.glm.html#module-adobo.glm">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="adobo.irlbpy.html">adobo.irlbpy package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#module-adobo.irlbpy.irlb">adobo.irlbpy.irlb module</a></li>
<li class="toctree-l2"><a class="reference internal" href="adobo.irlbpy.html#module-adobo.irlbpy">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-adobo.IO">
<span id="adobo-io-module"></span><h2>adobo.IO module<a class="headerlink" href="#module-adobo.IO" title="Permalink to this headline">¶</a></h2>
<div class="section" id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<p>Functions for reading and writing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.IO.export_data">
<code class="sig-prename descclassname">adobo.IO.</code><code class="sig-name descname">export_data</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">filename</em>, <em class="sig-param">norm='standard'</em>, <em class="sig-param">clust='leiden'</em>, <em class="sig-param">what='normalized'</em>, <em class="sig-param">transpose=False</em>, <em class="sig-param">sep='\t'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.IO.export_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Exports data to a text file, convenient for loading into other programs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>) – Output filename or path.</p></li>
<li><p><strong>norm</strong> (<cite>str</cite>) – Name of the normalisation. For example ‘standard’.</p></li>
<li><p><strong>clust</strong> (<cite>str</cite>) – Name of the clustering. For example: ‘leiden̈́’.</p></li>
<li><p><strong>what</strong> (<cite>{‘normalized’, ‘clusters’, ‘pca’, ‘tsne’, ‘umap’}</cite>) – What to export. Normalized data or PCA components.</p></li>
<li><p><strong>transpose</strong> (<cite>bool</cite>) – Transpose the data before writing it. Default: False</p></li>
<li><p><strong>sep</strong> (<cite>str</cite>) – A character or regular expression used to separate fields. Default: ”   “</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.IO.load_from_file">
<code class="sig-prename descclassname">adobo.IO.</code><code class="sig-name descname">load_from_file</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">sep='\\s'</em>, <em class="sig-param">header=True</em>, <em class="sig-param">desc='no desc set'</em>, <em class="sig-param">output_file=None</em>, <em class="sig-param">sparse=True</em>, <em class="sig-param">bundled=False</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.IO.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a gene expression matrix consisting of raw read counts</p>
<p class="rubric">Notes</p>
<p>The loaded gene expression matrix should not have been normalized. This function calls
<code class="xref py py-func docutils literal notranslate"><span class="pre">fread()</span></code> to read the data matrix file. Any additional arguments are
passed into it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<cite>str</cite>) – Path to the file containing input data. Should be a matrix where
columns are cells and rows are genes.</p></li>
<li><p><strong>sep</strong> (<cite>str</cite>) – A character or regular expression used to separate fields. Default: “\s”
(i.e. any white space character)</p></li>
<li><p><strong>header</strong> (<cite>bool</cite>) – If the data file has a header or not. Default: True</p></li>
<li><p><strong>desc</strong> (<cite>str</cite>) – A description of the data</p></li>
<li><p><strong>output_file</strong> (<cite>str</cite>) – An output filename used when calling <a class="reference internal" href="#adobo.data.dataset.save" title="adobo.data.dataset.save"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.data.dataset.save()</span></code></a>.</p></li>
<li><p><strong>sparse</strong> (<cite>bool</cite>) – Represent the data in a sparse data structure. Will save memory at the expense
of time. Default: True</p></li>
<li><p><strong>bundled</strong> (<cite>bool</cite>) – Use data installed by adobo. Default: False</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – To be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dataset class object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.bio">
<span id="adobo-bio-module"></span><h2>adobo.bio module<a class="headerlink" href="#module-adobo.bio" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Summary<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>Functions related to biology.</p>
<dl class="function">
<dt id="adobo.bio.cell_cycle_predict">
<code class="sig-prename descclassname">adobo.bio.</code><code class="sig-name descname">cell_cycle_predict</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">clf</em>, <em class="sig-param">tr_features</em>, <em class="sig-param">name=()</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.bio.cell_cycle_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts cell cycle phase</p>
<p class="rubric">Notes</p>
<p>The classifier is trained on mouse data, so it should _only_ be used on mouse data
unless it is trained on something else. Gene identifiers must use ensembl identifiers
(prefixed with ‘ENSMUSG’); pure gene symbols are not enough. Results are returned as
a column in the data frame <cite>meta_cells</cite> of the passed object. Does not return
probability scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>clf</strong> (<cite>sklearn.linear_model.SGDClassifier</cite>) – The classifier.</p></li>
<li><p><strong>tr_features</strong> (<cite>list</cite>) – Training features.</p></li>
<li><p><strong>name</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all available
normalizations will be used.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.bio.cell_cycle_train">
<code class="sig-prename descclassname">adobo.bio.</code><code class="sig-name descname">cell_cycle_train</code><span class="sig-paren">(</span><em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.bio.cell_cycle_train" title="Permalink to this definition">¶</a></dt>
<dd><p>Trains a cell cycle classifier using Stochastic Gradient Descent with data from
Buettner et al.</p>
<p class="rubric">Notes</p>
<p>Genes are selected from <a class="reference external" href="GO:0007049">GO:0007049</a></p>
<p>Does only need to be trained once; the second time it is serialized from disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets">1</span></dt>
<dd><p>Buettner et al. (2015) Computational analysis of cell-to-cell heterogeneity in
single-cell RNA-sequencing data reveals hidden subpopulations of cells. Nat
Biotech.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>sklearn.linear_model.SGDClassifier</cite> – A trained classifier.</p></li>
<li><p><cite>list</cite> – Containing training features.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.bio.cell_type_predict">
<code class="sig-prename descclassname">adobo.bio.</code><code class="sig-name descname">cell_type_predict</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">name=()</em>, <em class="sig-param">clustering=()</em>, <em class="sig-param">min_cluster_size=10</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.bio.cell_type_predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts cell types using the expression of marker genes</p>
<p class="rubric">Notes</p>
<p>Gene identifiers should be in symbol form, not ensembl identifiers, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>name</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all available
normalizations will be used.</p></li>
<li><p><strong>clustering</strong> (<cite>tuple</cite>, optional) – Specifies the clustering outcomes to work on.</p></li>
<li><p><strong>min_cluster_size</strong> (<cite>int</cite>) – Minimum number of cells per cluster; clusters smaller than this are ignored.
Default: 10</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.clustering">
<span id="adobo-clustering-module"></span><h2>adobo.clustering module<a class="headerlink" href="#module-adobo.clustering" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id3">
<h3>Summary<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>This module contains functions to cluster data.</p>
<dl class="function">
<dt id="adobo.clustering.generate">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">generate</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">k=10</em>, <em class="sig-param">name=None</em>, <em class="sig-param">distance='euclidean'</em>, <em class="sig-param">graph='snn'</em>, <em class="sig-param">clust_alg='leiden'</em>, <em class="sig-param">prune_snn=0.067</em>, <em class="sig-param">res=0.8</em>, <em class="sig-param">save_graph=True</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.generate" title="Permalink to this definition">¶</a></dt>
<dd><p>A wrapper function for generating single cell clusters from a shared nearest neighbor
graph with the Leiden algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>k</strong> (<cite>int</cite>) – Number of nearest neighbors. Default: 10</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None then the
function will be applied on all normalizations available.</p></li>
<li><p><strong>distance</strong> (<cite>str</cite>) – Distance metric to use. See here for valid choices: <a class="reference external" href="https://tinyurl.com/y4bckf7w">https://tinyurl.com/y4bckf7w</a>
Default: ‘euclidean’</p></li>
<li><p><strong>target</strong> (<cite>{‘irlb’, ‘svd’}</cite>) – The dimensionality reduction result to run on. Default: irlb</p></li>
<li><p><strong>graph</strong> (<cite>{‘snn’}</cite>) – Type of graph to generate. Only shared nearest neighbor (snn) supported at the
moment.</p></li>
<li><p><strong>clust_alg</strong> (<a href="#id4"><span class="problematic" id="id5">`</span></a>{‘leiden’, ‘walktrap’, ‘spinglass’, ‘multilevel’, ‘infomap’,) – ‘label_prop’, ‘leading_eigenvector’}`
Clustering algorithm to be used.</p></li>
<li><p><strong>prune_snn</strong> (<cite>float</cite>) – Threshold for pruning the SNN graph, i.e. the edges  with lower value (Jaccard
index) than this will be removed. Set to 0 to disable pruning. Increasing this
value will result in fewer edges in the graph. Default: 0.067</p></li>
<li><p><strong>res</strong> (<cite>float</cite>) – Resolution parameter for the Leiden algorithm _only_; change to modify cluster
resolution. Default: 0.8</p></li>
<li><p><strong>save_graph</strong> (<cite>bool</cite>) – To save the graph or not. Default: True</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id6"><span class="brackets">1</span></dt>
<dd><p>Yang et al. (2016) A Comparative Analysis of Community Detection Algorithms on
Artificial Networks. Scientific Reports</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dict containing cluster sizes (number of cells), only retx is set to True.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.clustering.igraph">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">igraph</code><span class="sig-paren">(</span><em class="sig-param">snn_graph</em>, <em class="sig-param">clust_alg</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.igraph" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs clustering functions within igraph</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snn_graph</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – Source and target nodes.</p></li>
<li><p><strong>clust_alg</strong> (<a href="#id7"><span class="problematic" id="id8">`</span></a>{‘walktrap’, ‘spinglass’, ‘multilevel’, ‘infomap’, ‘label_prop’,) – ‘leading_eigenvector’}`
Specifies the community detection algorithm.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id9"><span class="brackets">1</span></dt>
<dd><p>Pons &amp; Latapy (2006) Computing Communities in Large NetworksUsing Random Walks,
Journal of Graph Algorithms and Applications</p>
</dd>
<dt class="label" id="id10"><span class="brackets">2</span></dt>
<dd><p>Reichardt &amp; Bornholdt (2006) Statistical mechanics of community detection,
Physical Review E</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.clustering.knn">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">knn</code><span class="sig-paren">(</span><em class="sig-param">comp</em>, <em class="sig-param">k=10</em>, <em class="sig-param">distance='euclidean'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.knn" title="Permalink to this definition">¶</a></dt>
<dd><p>Nearest Neighbour Search. Finds the k number of near neighbours for each cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>comp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing PCA components.</p></li>
<li><p><strong>k</strong> (<cite>int</cite>) – Number of nearest neighbors. Default: 10</p></li>
<li><p><strong>target</strong> (<cite>{‘irlb’, ‘svd’}</cite>) – The dimensionality reduction result to run the NN search on. Default: irlb</p></li>
<li><p><strong>distance</strong> (<cite>str</cite>) – Distance metric to use. See here for valid choices: <a class="reference external" href="https://tinyurl.com/y4bckf7w">https://tinyurl.com/y4bckf7w</a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array containing indices.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.clustering.leiden">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">leiden</code><span class="sig-paren">(</span><em class="sig-param">snn_graph</em>, <em class="sig-param">res=0.8</em>, <em class="sig-param">seed=42</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.leiden" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs the Leiden algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>snn_graph</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – Source and target nodes.</p></li>
<li><p><strong>res</strong> (<cite>float</cite>) – Resolution parameter, change to modify cluster resolution. Default: 0.8</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id11"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://github.com/vtraag/leidenalg">https://github.com/vtraag/leidenalg</a></p>
</dd>
<dt class="label" id="id12"><span class="brackets">2</span></dt>
<dd><p>Traag et al. (2018) <a class="reference external" href="https://arxiv.org/abs/1810.08473">https://arxiv.org/abs/1810.08473</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.clustering.snn">
<code class="sig-prename descclassname">adobo.clustering.</code><code class="sig-name descname">snn</code><span class="sig-paren">(</span><em class="sig-param">nn_idx</em>, <em class="sig-param">k=10</em>, <em class="sig-param">prune_snn=0.067</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.clustering.snn" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a Shared Nearest Neighbor (SNN) graph</p>
<p class="rubric">Notes</p>
<p>Link weights are number of shared nearest neighbors. The sum of SNN similarities over
all KNNs is retrieved with linear algebra.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nn_idx</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code>) – Numpy array generated using knn()</p></li>
<li><p><strong>k</strong> (<cite>int</cite>) – Number of nearest neighbors. Default: 10</p></li>
<li><p><strong>prune_snn</strong> (<cite>float</cite>) – Threshold for pruning the SNN graph, i.e. the edges  with lower value (Jaccard
index) than this will be removed. Set to 0 to disable pruning. Increasing this
value will result in fewer edges in the graph. Default: 0.067</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id13"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="http://mlwiki.org/index.php/SNN_Clustering">http://mlwiki.org/index.php/SNN_Clustering</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.data">
<span id="adobo-data-module"></span><h2>adobo.data module<a class="headerlink" href="#module-adobo.data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id14">
<h3>Summary<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<p>This module contains a data storage class.</p>
<dl class="class">
<dt id="adobo.data.dataset">
<em class="property">class </em><code class="sig-prename descclassname">adobo.data.</code><code class="sig-name descname">dataset</code><span class="sig-paren">(</span><em class="sig-param">raw_mat</em>, <em class="sig-param">desc='no desc set'</em>, <em class="sig-param">output_file=None</em>, <em class="sig-param">input_file=None</em>, <em class="sig-param">sparse=True</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Storage container for raw, imputed and normalized data as well as analysis results.</p>
<dl class="attribute">
<dt id="adobo.data.dataset._assays">
<code class="sig-name descname">_assays</code><a class="headerlink" href="#adobo.data.dataset._assays" title="Permalink to this definition">¶</a></dt>
<dd><p>Holding information about what functions have been applied.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.count_data">
<code class="sig-name descname">count_data</code><a class="headerlink" href="#adobo.data.dataset.count_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw read count matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.imp_count_data">
<code class="sig-name descname">imp_count_data</code><a class="headerlink" href="#adobo.data.dataset.imp_count_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw data after imputing dropouts.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset._low_quality_cells">
<code class="sig-name descname">_low_quality_cells</code><a class="headerlink" href="#adobo.data.dataset._low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Low quality cells identified with <a class="reference internal" href="#adobo.preproc.find_low_quality_cells" title="adobo.preproc.find_low_quality_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adobo.preproc.find_low_quality_cells()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset._norm_data">
<code class="sig-name descname">_norm_data</code><a class="headerlink" href="#adobo.data.dataset._norm_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores all analysis results. A nested dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>dict</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.meta_cells">
<code class="sig-name descname">meta_cells</code><a class="headerlink" href="#adobo.data.dataset.meta_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>A data frame containing meta data for cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>pandas.DataFrame</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.meta_genes">
<code class="sig-name descname">meta_genes</code><a class="headerlink" href="#adobo.data.dataset.meta_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>A data frame containing meta data for genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>pandas.DataFrame</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.desc">
<code class="sig-name descname">desc</code><a class="headerlink" href="#adobo.data.dataset.desc" title="Permalink to this definition">¶</a></dt>
<dd><p>A string describing the dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.sparse">
<code class="sig-name descname">sparse</code><a class="headerlink" href="#adobo.data.dataset.sparse" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the data in a sparse data structure. Will save memory at the expense
of time. Default: True</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>bool</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.output_file">
<code class="sig-name descname">output_file</code><a class="headerlink" href="#adobo.data.dataset.output_file" title="Permalink to this definition">¶</a></dt>
<dd><p>A filename that will be used when calling save().</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite>, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.version">
<code class="sig-name descname">version</code><a class="headerlink" href="#adobo.data.dataset.version" title="Permalink to this definition">¶</a></dt>
<dd><p>The adobo package version used to create this data object.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.add_meta_data">
<code class="sig-name descname">add_meta_data</code><span class="sig-paren">(</span><em class="sig-param">axis</em>, <em class="sig-param">key</em>, <em class="sig-param">data</em>, <em class="sig-param">type_='cat'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.add_meta_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Add meta data to the adobo object</p>
<p class="rubric">Notes</p>
<p>Meta data can be added to cells or genes.</p>
<p>The parameter name ‘<a href="#id55"><span class="problematic" id="id56">type_</span></a>’ has an underscore to avoid conflict with Python’s
internal type keyword.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<cite>{‘cells’, ‘genes’}</cite>) – Are the data for cells or genes?</p></li>
<li><p><strong>key</strong> (<cite>str</cite>) – The variable name for your data. No whitespaces and special characters.</p></li>
<li><p><strong>data</strong> (<cite>numpy.ndarray</cite>, <cite>list</cite> or <cite>pandas.Series</cite>) – Data to add. Can be a basic Python <cite>list</cite>, a numpy array or a Pandas Series
with an index. If the data type is numpy array or list, then the length must
match the length of cells or genes. If the data type is a Pandas series, then
the length does not need to match as long as the index is there. Data can be
continuous or categorical and this must be specified with <cite>type_</cite>.</p></li>
<li><p><strong>type_</strong> (<cite>{‘cat’, ‘cont’}</cite>) – Specify if data are categorical or continuous. <cite>cat</cite> means categorical data
and <cite>cont</cite> means continuous data. Default: ‘cat’</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.assays">
<code class="sig-name descname">assays</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.assays" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays a basic summary of the dataset and what analyses have been performed on
it.</p>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.df_mem_usage">
<code class="sig-name descname">df_mem_usage</code><span class="sig-paren">(</span><em class="sig-param">var</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.df_mem_usage" title="Permalink to this definition">¶</a></dt>
<dd><p>Memory usage for a data frame in mega bytes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>var</strong> (<cite>str</cite>) – Variable name as a string.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Mega bytes used with two decimals.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><cite>float</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.get_assay">
<code class="sig-name descname">get_assay</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">lang=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.get_assay" title="Permalink to this definition">¶</a></dt>
<dd><p>Get info if a function has been applied.</p>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.is_normalized">
<code class="sig-name descname">is_normalized</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.is_normalized" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if normalized data can be found</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if it is normalized otherwise False.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>bool</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.low_quality_cells">
<em class="property">property </em><code class="sig-name descname">low_quality_cells</code><a class="headerlink" href="#adobo.data.dataset.low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.norm_data">
<em class="property">property </em><code class="sig-name descname">norm_data</code><a class="headerlink" href="#adobo.data.dataset.norm_data" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.print_dict">
<code class="sig-name descname">print_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.print_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">compress=True</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serializes the object</p>
<p class="rubric">Notes</p>
<p>This is a method so that it is not needed to memorize the filename, instead the
filename was already specified when the object was created with the
<cite>output_file</cite> parameter. Load the object data with <cite>joblib.load</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>compress</strong> (<cite>bool</cite>) – Save with data compression or not. Default: True</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="adobo.data.dataset.set_assay">
<code class="sig-name descname">set_assay</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">key=1</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset.set_assay" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the assay that was applied.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.de">
<span id="adobo-de-module"></span><h2>adobo.de module<a class="headerlink" href="#module-adobo.de" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id15">
<h3>Summary<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<p>Functions for differential expression.</p>
<dl class="function">
<dt id="adobo.de.combine_tests">
<code class="sig-prename descclassname">adobo.de.</code><code class="sig-name descname">combine_tests</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">normalization=()</em>, <em class="sig-param">clustering=()</em>, <em class="sig-param">method='simes'</em>, <em class="sig-param">min_cluster_size=10</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.de.combine_tests" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses Simes’ method for combining p-values</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>normalization</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all available
normalizations will be used.</p></li>
<li><p><strong>clustering</strong> (<cite>tuple</cite>, optional) – Specifies the clustering outcomes to work on.</p></li>
<li><p><strong>method</strong> (<cite>{‘simes’}</cite>) – Method for combining p-values. Only Simes’ is implemented.</p></li>
<li><p><strong>min_cluster_size</strong> (<cite>int</cite>) – Minimum number of cells per cluster (clusters smaller than this are ignored).
Default: 10</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id16"><span class="brackets">1</span></dt>
<dd><p>Simes, R. J. (1986). An improved Bonferroni procedure for multiple tests of
significance. Biometrika, 73(3):751-754.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.de.linear_model">
<code class="sig-prename descclassname">adobo.de.</code><code class="sig-name descname">linear_model</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">normalization=()</em>, <em class="sig-param">clustering=()</em>, <em class="sig-param">direction='up'</em>, <em class="sig-param">min_cluster_size=10</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.de.linear_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs differential expression analysis between clusters using a linear model
and t-statistics</p>
<p class="rubric">Notes</p>
<p>Finds differentially expressed (DE) genes between clusters by fitting a linear model
(LM) to gene expression (response variables) and clusters (explanatory variables) and
performs pairwise t-tests for significance. Model coefficients are estimated via the
ordinary least squares method and p-values are calculated using t-statistics. One
benefit of using LM for DE is that computations are vectorized and therefore very fast.</p>
<p>The ideas behind using LM to explore DE have been extensively covered in the
limma R package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>normalization</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all available
normalizations will be used.</p></li>
<li><p><strong>clustering</strong> (<cite>tuple</cite>, optional) – Specifies the clustering outcomes to work on.</p></li>
<li><p><strong>direction</strong> (<cite>{‘up’, ‘down’, ‘any’}</cite>) – Can be ‘any’ for any direction ‘up’ for up-regulated and ‘down’ for
down-regulated. Normally we want to find genes being upregulated in cluster A
compared with cluster B. Default: up</p></li>
<li><p><strong>min_cluster_size</strong> (<cite>int</cite>) – Minimum number of cells per cluster (clusters smaller than this are ignored).
Default: 10</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id17"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">https://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf</a></p>
</dd>
<dt class="label" id="id18"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://newonlinecourses.science.psu.edu/stat555/node/12/">https://newonlinecourses.science.psu.edu/stat555/node/12/</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.dr">
<span id="adobo-dr-module"></span><h2>adobo.dr module<a class="headerlink" href="#module-adobo.dr" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id19">
<h3>Summary<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<p>Functions for dimensional reduction.</p>
<dl class="function">
<dt id="adobo.dr.force_graph">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">force_graph</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">name=()</em>, <em class="sig-param">iterations=1000</em>, <em class="sig-param">edgeWeightInfluence=1.0</em>, <em class="sig-param">jitterTolerance=1.0</em>, <em class="sig-param">barnesHutOptimize=True</em>, <em class="sig-param">scalingRatio=2.0</em>, <em class="sig-param">gravity=1.0</em>, <em class="sig-param">strongGravityMode=False</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.force_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a force-directed graph</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. Default is to run on all.</p></li>
<li><p><strong>iterations</strong> (<cite>int</cite>) – Number of iterations. Default: 1000</p></li>
<li><p><strong>edgeWeightInfluence</strong> (<cite>float</cite>) – How much influence to edge weights. 0 is no influence and 1 is normal.
Default: 1.0</p></li>
<li><p><strong>jitterTolerance</strong> (<cite>float</cite>) – Amount swing. Lower gives less speed and more precision. Default: 1.0</p></li>
<li><p><strong>barnesHutOptimize</strong> (<cite>bool</cite>) – Run Barnes Hut optimization. Default: True</p></li>
<li><p><strong>scalingRatio</strong> (<cite>float</cite>) – Amount of repulsion, higher values make a more sparse graph. Default: 2.0</p></li>
<li><p><strong>gravity</strong> (<cite>float</cite>) – Attracts nodes to the center. Prevents islands from drifting away. Default: 1.0</p></li>
<li><p><strong>strongGravityMode</strong> (<cite>bool</cite>) – A stronger gravity view. Default: False</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id20"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://en.wikipedia.org/wiki/Force-directed_graph_drawing">https://en.wikipedia.org/wiki/Force-directed_graph_drawing</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.irlb">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">irlb</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ncomp=75</em>, <em class="sig-param">seed=None</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.irlb" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncated SVD by implicitly restarted Lanczos bidiagonalization</p>
<p class="rubric">Notes</p>
<p>The augmented implicitly restarted Lanczos bidiagonalization algorithm (IRLBA) finds
a few approximate largest singular values and corresponding singular vectors using a
method of Baglama and Reichel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return. Default: 75</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility. Default: None</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id21"><span class="brackets">1</span></dt>
<dd><p>Baglama et al (2005) Augmented Implicitly Restarted Lanczos Bidiagonalization
Methods SIAM Journal on Scientific Computing</p>
</dd>
<dt class="label" id="id22"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://github.com/bwlewis/irlbpy">https://github.com/bwlewis/irlbpy</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the components (columns).</p></li>
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the contributions of every gene (rows).</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.pca">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">pca</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='irlb'</em>, <em class="sig-param">name=None</em>, <em class="sig-param">ncomp=75</em>, <em class="sig-param">allgenes=False</em>, <em class="sig-param">scale=True</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">seed=None</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Principal Component Analysis</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual normalization functions, which can also be
called directly. Scaling of the data is achieved by setting scale=True (default),
which will center (subtract the column mean) and scale columns (divide by their
standard deviation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘irlb’, ‘svd’}</cite>) – Method to use for PCA. This does not matter much. Default: irlb</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None then the
function will be applied on all normalizations available.</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return. Default: 75</p></li>
<li><p><strong>allgenes</strong> (<cite>bool</cite>) – Use all genes instead of only HVG. Default: False</p></li>
<li><p><strong>scale</strong> (<cite>bool</cite>) – Scales input data prior to PCA. Default: True</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be noisy or not. Default: False</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility (only irlb). Default: None</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id23"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://en.wikipedia.org/wiki/Principal_component_analysis">https://en.wikipedia.org/wiki/Principal_component_analysis</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>Nothing. Modifies the passed object. Results are stored in two dictonaries in the</em></p></li>
<li><p><strong>passed object</strong> (<cite>dr</cite> (containing the components) and <cite>dr_gene_contr</cite> (containing)</p></li>
<li><p><em>gene contributions to each component)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.svd">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">svd</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ncomp=75</em>, <em class="sig-param">only_sdev=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.svd" title="Permalink to this definition">¶</a></dt>
<dd><p>Principal component analysis via singular value decomposition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data. Preferrably this
should be a subset of the normalized gene expression matrix containing highly
variable genes.</p></li>
<li><p><strong>ncomp</strong> (<cite>int</cite>) – Number of components to return. Default: 75</p></li>
<li><p><strong>only_sdev</strong> (<cite>bool</cite>) – Only return the standard deviation of the components. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id24"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://tinyurl.com/yyt6df5x">https://tinyurl.com/yyt6df5x</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the components (columns). Only if
only_sdev=False.</p></li>
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing the contributions of every gene (rows).
Only if only_sdev=False.</p></li>
<li><p><cite>pd.DataFrame</cite> – A py:class:<cite>pandas.DataFrame</cite> containing standard deviations of components. Only
if only_sdev is set to True.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.tsne">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">tsne</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">run_on_PCA=True</em>, <em class="sig-param">name=None</em>, <em class="sig-param">perplexity=30</em>, <em class="sig-param">n_iter=2000</em>, <em class="sig-param">seed=None</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.tsne" title="Permalink to this definition">¶</a></dt>
<dd><p>Projects data to a two dimensional space using the tSNE algorithm.</p>
<p class="rubric">Notes</p>
<p>It is recommended to perform this function on data in PCA space. This function calls
<code class="xref py py-func docutils literal notranslate"><span class="pre">sklearn.manifold.TSNE()</span></code>, and any additional parameters will be passed to it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>run_on_PCA</strong> (<cite>bool</cite>) – To run tSNE on PCA components or not. If False then runs on the entire normalized
gene expression matrix. Default: True</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None then the
function will be applied on all normalizations available.</p></li>
<li><p><strong>perplexity</strong> (<cite>float</cite>) – From [1]: The perplexity is related to the number of nearest neighbors that
is used in other manifold learning algorithms. Larger datasets usually require
a larger perplexity. Consider selecting a value between 5 and 50. Different
values can result in significanlty different results. Default: 30</p></li>
<li><p><strong>n_iter</strong> (<cite>int</cite>) – Number of iterations. Default: 2000</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility. Default: None</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id25"><span class="brackets">1</span></dt>
<dd><p>van der Maaten, L.J.P.; Hinton, G.E. Visualizing High-Dimensional Data
Using t-SNE. Journal of Machine Learning Research 9:2579-2605, 2008.</p>
</dd>
<dt class="label" id="id26"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html">https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.dr.umap">
<code class="sig-prename descclassname">adobo.dr.</code><code class="sig-name descname">umap</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">run_on_PCA=True</em>, <em class="sig-param">name=None</em>, <em class="sig-param">n_neighbors=15</em>, <em class="sig-param">distance='euclidean'</em>, <em class="sig-param">n_epochs=None</em>, <em class="sig-param">learning_rate=1.0</em>, <em class="sig-param">min_dist=0.1</em>, <em class="sig-param">spread=1.0</em>, <em class="sig-param">seed=None</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.dr.umap" title="Permalink to this definition">¶</a></dt>
<dd><p>Projects data to a low-dimensional space using the Uniform Manifold Approximation
and Projection (UMAP) algorithm</p>
<p class="rubric">Notes</p>
<p>UMAP is a non-linear data reduction algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>run_on_PCA</strong> (<cite>bool</cite>) – To run tSNE on PCA components or not. If False then runs on the entire normalized
gene expression matrix. Default: True</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None then the
function will be applied on all normalizations available.</p></li>
<li><p><strong>n_neighbors</strong> (<cite>int</cite>) – The size of local neighborhood (in terms of number of neighboring sample points)
used for manifold approximation. Larger values result in more global views of the
manifold, while smaller values result in more local data being preserved.
In general values should be in the range 2 to 100. Default: 15</p></li>
<li><p><strong>distance</strong> (<cite>str</cite>) – The metric to use to compute distances in high dimensional space.
Default: ‘euclidean’</p></li>
<li><p><strong>n_epochs</strong> (<cite>int</cite>) – The number of training epochs to be used in optimizing the low dimensional
embedding. Larger values result in more accurate embeddings. If None is specified
a value will be selected based on the size of the input dataset (200 for large
datasets, 500 for small). Default: None</p></li>
<li><p><strong>learning_rate</strong> (<cite>float</cite>) – The initial learning rate for the embedding optimization. Default: 1.0</p></li>
<li><p><strong>min_dist</strong> (<cite>float</cite>) – The effective minimum distance between embedded points. Default: 0.1</p></li>
<li><p><strong>spread</strong> (<cite>float</cite>) – The effective scale of embedded points. Default: 1.0</p></li>
<li><p><strong>seed</strong> (<cite>int</cite>) – For reproducibility. Default: None</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id27"><span class="brackets">1</span></dt>
<dd><p>McInnes L, Healy J, Melville J (2018) UMAP: Uniform Manifold Approximation and
Projection for Dimension Reduction, <a class="reference external" href="https://arxiv.org/abs/1802.03426">https://arxiv.org/abs/1802.03426</a></p>
</dd>
<dt class="label" id="id28"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://github.com/lmcinnes/umap">https://github.com/lmcinnes/umap</a></p>
</dd>
<dt class="label" id="id29"><span class="brackets">3</span></dt>
<dd><p><a class="reference external" href="https://umap-learn.readthedocs.io/en/latest/">https://umap-learn.readthedocs.io/en/latest/</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.hvg">
<span id="adobo-hvg-module"></span><h2>adobo.hvg module<a class="headerlink" href="#module-adobo.hvg" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id30">
<h3>Summary<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<p>Functions for detection of highly variable genes.</p>
<dl class="function">
<dt id="adobo.hvg.brennecke">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">brennecke</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log</em>, <em class="sig-param">ercc=Empty DataFrame Columns: [] Index: []</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">minBiolDisp=0.5</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.brennecke" title="Permalink to this definition">¶</a></dt>
<dd><p>Implements the method of Brennecke et al. (2013) to identify highly variable genes</p>
<p class="rubric">Notes</p>
<p>Fits data using GLM with Fisher Scoring. GLM code copied from (credits to &#64;madrury
for this code): <a class="reference external" href="https://github.com/madrury/py-glm">https://github.com/madrury/py-glm</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>log</strong> (<cite>bool</cite>) – If normalized data were log transformed or not.</p></li>
<li><p><strong>ercc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized ercc spikes.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>minBiolDisp</strong> (<cite>float</cite>) – Minimum percentage of variance due to biological factors.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id31"><span class="brackets">1</span></dt>
<dd><p>Brennecke et al. (2013) Nature Methods <a class="reference external" href="https://doi.org/10.1038/nmeth.2645">https://doi.org/10.1038/nmeth.2645</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.chen2016">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">chen2016</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.chen2016" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from Chen (2016) to identify highly variable
genes.</p>
<p class="rubric">Notes</p>
<p>Expression counts should be normalized and on a log scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>log</strong> (<cite>bool</cite>) – If normalized data were log transformed or not.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id32"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-2897-6">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-2897-6</a></p>
</dd>
<dt class="label" id="id33"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://github.com/hillas/scVEGs/blob/master/scVEGs.r">https://github.com/hillas/scVEGs/blob/master/scVEGs.r</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.find_hvg">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">find_hvg</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='seurat'</em>, <em class="sig-param">normalization=None</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.find_hvg" title="Permalink to this definition">¶</a></dt>
<dd><p>Finding highly variable genes</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual HVG functions, which can also be called
directly.</p>
<p>The method ‘brennecke’ should not be applied on ‘fqn’ normalized data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘seurat’, ‘brennecke’, ‘scran’, ‘chen2016’, ‘mm’}</cite>) – Specifies the method to be used.</p></li>
<li><p><strong>normalization</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None then the
function will be applied on all normalizations available.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of genes to return.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate threshold for significant genes applied to those methods
that use it (brennecke, chen2016, mm). Note that the number of returned genes
might be fewer than specified by <cite>ngenes</cite> because of FDR consideration.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id34"><span class="brackets">1</span></dt>
<dd><p>Yip et al. (2018) Briefings in Bioinformatics
<a class="reference external" href="https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby011/4898116">https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby011/4898116</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.mm">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">mm</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">log</em>, <em class="sig-param">fdr=0.1</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.mm" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from Andrews (2018).</p>
<p class="rubric">Notes</p>
<p>Input should be normalized but nog log’ed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized counts.</p></li>
<li><p><strong>fdr</strong> (<cite>float</cite>) – False Discovery Rate considered significant.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id35"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty1044">https://doi.org/10.1093/bioinformatics/bty1044</a></p>
</dd>
<dt class="label" id="id36"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://github.com/tallulandrews/M3Drop">https://github.com/tallulandrews/M3Drop</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.scran">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">scran</code><span class="sig-paren">(</span><em class="sig-param">data_norm</em>, <em class="sig-param">ercc</em>, <em class="sig-param">log</em>, <em class="sig-param">ngenes=1000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.scran" title="Permalink to this definition">¶</a></dt>
<dd><p>This function implements the approach from the scran R package</p>
<p class="rubric">Notes</p>
<p>Expression counts should be normalized and on a log scale.</p>
<p>Outline of the steps:</p>
<ol class="arabic simple">
<li><p>fits a polynomial regression model to mean and variance of the technical genes</p></li>
<li><p>decomposes the total variance of the biological genes by subtracting the
technical variance predicted by the fit</p></li>
<li><p>sort based on biological variance</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_norm</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized gene expression data.</p></li>
<li><p><strong>ercc</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing normalized ercc spikes.</p></li>
<li><p><strong>log</strong> (<cite>bool</cite>) – If normalized data were log transformed or not.</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id37"><span class="brackets">1</span></dt>
<dd><p>Lun ATL, McCarthy DJ, Marioni JC (2016). “A step-by-step workflow for low-level
analysis of single-cell RNA-seq data with Bioconductor.” F1000Research,
<a class="reference external" href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.hvg.seurat">
<code class="sig-prename descclassname">adobo.hvg.</code><code class="sig-name descname">seurat</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">ngenes=1000</em>, <em class="sig-param">num_bins=20</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.hvg.seurat" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves a list of highly variable genes using Seurat’s strategy</p>
<p class="rubric">Notes</p>
<p>The function bins the genes according to average expression, then calculates
dispersion for each bin as variance to mean ratio. Within each bin, Z-scores are
calculated and returned. Z-scores are ranked and the top 1000 are selected. Input data
should be normalized first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>ngenes</strong> (<cite>int</cite>) – Number of top highly variable genes to return.</p></li>
<li><p><strong>num_bins</strong> (<cite>int</cite>) – Number of bins to use.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id38"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://cran.r-project.org/web/packages/Seurat/index.html">https://cran.r-project.org/web/packages/Seurat/index.html</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list containing highly variable genes.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.normalize">
<span id="adobo-normalize-module"></span><h2>adobo.normalize module<a class="headerlink" href="#module-adobo.normalize" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id39">
<h3>Summary<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<p>This module contains functions to normalize raw read counts.</p>
<dl class="function">
<dt id="adobo.normalize.clr">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">clr</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">axis='genes'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.clr" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs centered log ratio normalization similar to Seurat</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>axis</strong> (<em>{'genes'</em><em>, </em><em>'cells'}</em>) – Normalize over genes or cells. Default: ‘genes’</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id40"><span class="brackets">1</span></dt>
<dd><p>Hafemeister et al. (2019) <a class="reference external" href="https://www.biorxiv.org/content/10.1101/576827v1">https://www.biorxiv.org/content/10.1101/576827v1</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;clr&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.fqn">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">fqn</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.fqn" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs full quantile normalization (FQN)</p>
<p class="rubric">Notes</p>
<p>FQN was shown to perform well on single cell data [1] and was a popular
normalization scheme for microarray data. The present function does not handle ties
well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id41"><span class="brackets">1</span></dt>
<dd><p>Bolstad et al. (2003) Bioinformatics
<a class="reference external" href="https://academic.oup.com/bioinformatics/article/19/2/185/372664">https://academic.oup.com/bioinformatics/article/19/2/185/372664</a></p>
</dd>
<dt class="label" id="id42"><span class="brackets">2</span></dt>
<dd><p>Cole et al. (2019) Cell Systems
<a class="reference external" href="https://www.biorxiv.org/content/10.1101/235382v2">https://www.biorxiv.org/content/10.1101/235382v2</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;fqn&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.norm">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">norm</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='standard'</em>, <em class="sig-param">name=None</em>, <em class="sig-param">use_imputed=False</em>, <em class="sig-param">log=True</em>, <em class="sig-param">log_func=&lt;ufunc 'log2'&gt;</em>, <em class="sig-param">small_const=1</em>, <em class="sig-param">remove_low_qual=True</em>, <em class="sig-param">gene_lengths=None</em>, <em class="sig-param">scaling_factor=10000</em>, <em class="sig-param">axis='genes'</em>, <em class="sig-param">retx=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalizes gene expression data</p>
<p class="rubric">Notes</p>
<p>A wrapper function around the individual normalization functions, which can also be
called directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<cite>{‘standard’, ‘rpkm’, ‘fqn’, ‘clr’, ‘vsn’}</cite>) – Specifies the method to use. <cite>standard</cite> refers to the simplest normalization
strategy involving scaling genes by total number of reads per cell. <cite>rpkm</cite>
performs RPKM normalization and requires the <cite>gene_lengths</cite> parameter to be set.
<cite>fqn</cite> performs a full-quantile normalization. <cite>clr</cite> performs centered log ratio
normalization. <cite>vsn</cite> performs a variance stabilizing normalization.
Default: standard</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – A choosen name for the normalization. It is used for storing and retrieving this
normalization for plotting later. If <cite>None</cite> or an empty string, then it is set to
the value of <cite>method</cite>.</p></li>
<li><p><strong>use_imputed</strong> (<cite>bool</cite>) – Use imputed data. If set to True, then <a class="reference internal" href="#adobo.preproc.impute" title="adobo.preproc.impute"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.preproc.impute()</span></code></a> must have been
run previously. Default: False</p></li>
<li><p><strong>log</strong> (<cite>bool</cite>) – Perform log transformation. Default: True</p></li>
<li><p><strong>log_func</strong> (<cite>numpy.func</cite>) – Logarithmic function to use. For example: np.log2, np.log1p, np.log10, etc.
Default: np.log2</p></li>
<li><p><strong>small_const</strong> (<cite>float</cite>) – A small constant to add to expression values to avoid log’ing genes with zero
expression. Default: 1</p></li>
<li><p><strong>remove_low_qual</strong> (<cite>bool</cite>) – Remove low quality cells and uninformative genes identified by prior steps.
Default: True</p></li>
<li><p><strong>gene_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> or <cite>str</cite>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> containing the gene lengths in base pairs and gene names
set as index. The names must match the gene names used in <cite>data</cite> (the order does
not need to match and any symbols not found in the data will be discarded).
Normally gene lengths should be the combined length of exons for every gene. If
gene_lengths is a <cite>str</cite> then it is taken as a filename and loaded; first column is
gene names and second column is the length, field separator is one space.
<cite>gene_lengths</cite> needs to be set _only_ if method=’rpkm’. Default: None</p></li>
<li><p><strong>scaling_factor</strong> (<cite>int</cite>) – Scaling factor used to multiply the scaled counts with. Only used for
<cite>method=”depth”</cite>. Default: 10000</p></li>
<li><p><strong>axis</strong> (<em>{'genes'</em><em>, </em><em>'cells'}</em>) – Only applicable when <cite>method=”clr”</cite>, defines the axis to normalize across.
Default: ‘genes’</p></li>
<li><p><strong>retx</strong> (<cite>bool</cite>) – Return the normalized data as well. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id43"><span class="brackets">1</span></dt>
<dd><p>Cole et al. (2019) Cell Systems
<a class="reference external" href="https://www.biorxiv.org/content/10.1101/235382v2">https://www.biorxiv.org/content/10.1101/235382v2</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">exp</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.rpkm">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">rpkm</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">gene_lengths</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.rpkm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize expression values as RPKM</p>
<p class="rubric">Notes</p>
<p>This method should be used if you need to adjust for gene length, such as in
a SMART-Seq2 protocol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>gene_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> or <cite>str</cite>) – Should contain the gene lengths in base pairs and gene names set as index. The
names must match the gene names used in <cite>data</cite>. Normally gene lengths should be
the combined length of exons for every gene. If gene_lengths is a <cite>str</cite> then it is
taken as a file path and loads it; first column is gene names and second column is
the length, field separator is one space; an alternative format is a single column
of combined exon lengths where the total number of rows matches the number of rows
in the raw read counts matrix and with the same order.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id44"><span class="brackets">1</span></dt>
<dd><p>Conesa et al. (2016) Genome Biology
<a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.standard">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">standard</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">scaling_factor=10000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.standard" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs a standard normalization by scaling with the total read depth per cell and
then multiplying with a scaling factor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>scaling_factor</strong> (<cite>int</cite>) – Scaling factor used to multiply the scaled counts with. Default: 10000</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id45"><span class="brackets">1</span></dt>
<dd><p>Evans et al. (2018) Briefings in Bioinformatics
<a class="reference external" href="https://academic.oup.com/bib/article/19/5/776/3056951">https://academic.oup.com/bib/article/19/5/776/3056951</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;standard&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.vsn">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">vsn</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">min_cells=5</em>, <em class="sig-param">gmean_eps=1</em>, <em class="sig-param">n_genes=2000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.vsn" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs variance stabilizing normalization based on a negative binomial regression
model with regularized parameters</p>
<p class="rubric">Notes</p>
<p>Use only with UMI counts. Adopts a subset of the functionality of <cite>vst</cite> in the R
package <cite>sctransform</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes, columns=cells).</p></li>
<li><p><strong>min_cells</strong> (<cite>int</cite>) – Minimum number of cells expressing a gene for the gene to be used. Default: 10</p></li>
<li><p><strong>gmean_eps</strong> (<cite>float</cite>) – A small constant to avoid log(0)=-Inf. Default: 1</p></li>
<li><p><strong>n_genes</strong> (<cite>int</cite>) – Number of genes to use when estimating parameters. Default: 2000</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id46"><span class="brackets">1</span></dt>
<dd><p><a class="reference external" href="https://cran.r-project.org/web/packages/sctransform/index.html">https://cran.r-project.org/web/packages/sctransform/index.html</a></p>
</dd>
<dt class="label" id="id47"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/576827v1">https://www.biorxiv.org/content/10.1101/576827v1</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;vsn&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A data matrix with adjusted counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.plotting">
<span id="adobo-plotting-module"></span><h2>adobo.plotting module<a class="headerlink" href="#module-adobo.plotting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id48">
<h3>Summary<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h3>
<p>Functions for plotting scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.plotting.cell_viz">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">cell_viz</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">reduction='tsne'</em>, <em class="sig-param">normalization=()</em>, <em class="sig-param">clustering=()</em>, <em class="sig-param">metadata=()</em>, <em class="sig-param">genes=()</em>, <em class="sig-param">edges=False</em>, <em class="sig-param">cell_types=False</em>, <em class="sig-param">trajectory=None</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">marker_size=0.8</em>, <em class="sig-param">font_size=8</em>, <em class="sig-param">colors='adobo'</em>, <em class="sig-param">title=None</em>, <em class="sig-param">legend=True</em>, <em class="sig-param">legend_marker_scale=10</em>, <em class="sig-param">legend_position=(1</em>, <em class="sig-param">1)</em>, <em class="sig-param">min_cluster_size=10</em>, <em class="sig-param">fig_size=(10</em>, <em class="sig-param">10)</em>, <em class="sig-param">margins=None</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.cell_viz" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a 2d scatter plot from an embedding</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>reduction</strong> (<cite>{‘tsne’, ‘umap’, ‘pca’, ‘force_graph’}</cite>) – The dimensional reduction to use. Default: tsne</p></li>
<li><p><strong>normalization</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all
available normalizations will be used.</p></li>
<li><p><strong>clustering</strong> (<cite>tuple</cite>, optional) – Specifies the clustering outcomes to plot.</p></li>
<li><p><strong>metadata</strong> (<cite>tuple</cite>, optional) – Specifies the metadata variables to plot.</p></li>
<li><p><strong>genes</strong> (<cite>tuple</cite>, optional) – Specifies genes to plot.</p></li>
<li><p><strong>edges</strong> (<cite>bool</cite>) – Draw edges (only applicable if reduction=’force_graph’). Default: False</p></li>
<li><p><strong>cell_types</strong> (<cite>bool</cite>) – Print cell type predictions, applicable if <a class="reference internal" href="#adobo.bio.cell_type_predict" title="adobo.bio.cell_type_predict"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.bio.cell_type_predict()</span></code></a>
has been run. Default: False</p></li>
<li><p><strong>trajectory</strong> (<cite>str</cite>, optional) – The trajectory to plot. For example ‘slingshot’. Default: None</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Name of an output file instead of showing on screen.</p></li>
<li><p><strong>marker_size</strong> (<cite>float</cite>) – The size of the markers.</p></li>
<li><p><strong>font_size</strong> (<cite>float</cite>) – Font size. Default: 8</p></li>
<li><p><strong>colors</strong> (<cite>{‘default’, ‘random’}</cite> or <cite>list</cite>) – Can be: (i) “adobo” or “random”; or (ii) a <cite>list</cite> of colors with the
same length as the number of factors. If colors is set to “adobo”, then
colors are retrieved from <code class="xref py py-attr docutils literal notranslate"><span class="pre">adobo._constants.CLUSTER_COLORS_DEFAULT</span></code>
(but if the number of clusters exceed 50, then random colors will be
used). Default: adobo</p></li>
<li><p><strong>title</strong> (<cite>str</cite>) – Title of the plot. By default the title is set to the reduction technique.</p></li>
<li><p><strong>legend</strong> (<cite>bool</cite>) – Add legend or not. Default: True</p></li>
<li><p><strong>legend_marker_scale</strong> (<cite>int</cite>) – Scale the markers in the legend. Default: 10</p></li>
<li><p><strong>legend_position</strong> (<cite>tuple</cite>) – A tuple of length two describing the position of the legend. Default: (1,1)</p></li>
<li><p><strong>min_cluster_size</strong> (<cite>int</cite>) – Can be used to prevent clusters below a certain number of cells to be
plotted. Default: 10</p></li>
<li><p><strong>fig_size</strong> (<cite>tuple</cite>) – Figure size in inches. Default: (10, 10)</p></li>
<li><p><strong>margins</strong> (<cite>dict</cite>) – Can be used to adjust margins. Should be a dict with one or more of the
keys: ‘left’, ‘bottom’, ‘right’, ‘top’, ‘wspace’, ‘hspace’. Set
verbose=True to figure out the present values. Default: None</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.overall">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">overall</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">what='reads'</em>, <em class="sig-param">how='histogram'</em>, <em class="sig-param">bin_size=100</em>, <em class="sig-param">cut_off=None</em>, <em class="sig-param">color='#E69F00'</em>, <em class="sig-param">title=None</em>, <em class="sig-param">filename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.overall" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a plot of read counts per cell or expressed genes per cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>what</strong> (<cite>{‘reads’, ‘genes’}</cite>) – If ‘reads’ then plots the number of reads per cell. If ‘genes’, then
plots the number of expressed genes per cell. Default: ‘reads’</p></li>
<li><p><strong>how</strong> (<cite>{‘histogram’, ‘boxplot’, ‘barplot’, ‘violin’}</cite>) – Type of plot to generate. Default: ‘histogram’</p></li>
<li><p><strong>bin_size</strong> (<cite>int</cite>) – If <cite>how</cite> is a histogram, then this is the bin size. Default: 100</p></li>
<li><p><strong>cut_off</strong> (<cite>int</cite>) – Set a cut off for genes or reads by drawing a red line and print the number
of cells over and under the cut off. Only valid if how=’histogram’.
Default: None</p></li>
<li><p><strong>color</strong> (<cite>str</cite>) – Color of the plot. Default: ‘#E69F00’</p></li>
<li><p><strong>title</strong> (<cite>str</cite>) – Change the default title of the plot. Default: None</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file instead of showing it on the screen. Default: None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.overall_scatter">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">overall_scatter</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">color='#E69F00'</em>, <em class="sig-param">title=None</em>, <em class="sig-param">filename=None</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.overall_scatter" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Generates a scatter plot showing the total number of reads on one axis</dt><dd><p>and the number of detected genes on the other axis</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>color</strong> (<cite>str</cite>) – Color of the plot. Default: ‘#E69F00’</p></li>
<li><p><strong>title</strong> (<cite>str</cite>) – Title of the plot. Default: None</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file instead of showing it on the screen. Default: None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.pca_contributors">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">pca_contributors</code><span class="sig-paren">(</span><em class="sig-param">obj, name=None, dim=[0, 1, 2], top=10, color='#fcc603', fontsize=6, fig_size=(10, 5), filename=None, verbose=False, **args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.pca_contributors" title="Permalink to this definition">¶</a></dt>
<dd><p>Examine the top contributing genes to each PCA component</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Genes are ranked by their absolute value. Additional parameters are passed
into <code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.savefig()</span></code>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>name</strong> (<cite>str</cite>) – The name of the normalization to operate on. If this is empty or None
then the function will be applied on all normalizations available.</p></li>
<li><p><strong>dim</strong> (<cite>list</cite>) – A list of indices specifying components to plot. First component has
index zero.</p></li>
<li><p><strong>top</strong> (<cite>int</cite>) – Specifies the number of top scoring genes to include. Default: 10</p></li>
<li><p><strong>color</strong> (<cite>str</cite>) – Color of the bars. As a string or hex code. Default: “#fcc603”</p></li>
<li><p><strong>fontsize</strong> (<cite>int</cite>) – Specifies font size. Default: 6</p></li>
<li><p><strong>fig_size</strong> (<cite>tuple</cite>) – Figure size in inches. Default: (10, 10)</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write to a file instead of showing the plot on screen.</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.pca_elbow">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">pca_elbow</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">name=()</em>, <em class="sig-param">comp_max=200</em>, <em class="sig-param">all_genes=False</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">font_size=8</em>, <em class="sig-param">fig_size=(10</em>, <em class="sig-param">10)</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">**args</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.pca_elbow" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a PCA elbow plot</p>
<p class="rubric">Notes</p>
<p>Can be used for determining the number of principal components to include. PCA is
based on singular value decomposition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>name</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all available
normalizations will be used.</p></li>
<li><p><strong>comp_max</strong> (<cite>int</cite>) – Maximum number of components to include.</p></li>
<li><p><strong>all_genes</strong> (<cite>bool</cite>) – Run on all genes, i.e. not only highly variable genes. Default: False</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Name of an output file instead of showing on screen.</p></li>
<li><p><strong>font_size</strong> (<cite>float</cite>) – Font size. Default: 8</p></li>
<li><p><strong>fig_size</strong> (<cite>tuple</cite>) – Figure size in inches. Default: (10, 10)</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: True</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.preproc">
<span id="adobo-preproc-module"></span><h2>adobo.preproc module<a class="headerlink" href="#module-adobo.preproc" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id49">
<h3>Summary<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<p>Functions for pre-processing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.preproc.find_ercc">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_ercc</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">ercc_pattern='^ERCC[_-]\\S+$'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_ercc" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag ERCC spikes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>ercc_pattern</strong> (<cite>str</cite>, optional) – A regular expression matching ercc gene symbols. Default: “ercc[_-]S+$”</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of detected ercc spikes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.find_low_quality_cells">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_low_quality_cells</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">rRNA_genes</em>, <em class="sig-param">sd_thres=3</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Statistical detection of low quality cells using Mahalanobis distances</p>
<p class="rubric">Notes</p>
<p>Mahalanobis distances are computed from five quality metrics. A robust estimate of
covariance is used in the Mahalanobis function. Cells with Mahalanobis distances of
three standard deviations from the mean are by default considered outliers.
The five metrics are:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>log-transformed number of molecules detected</p></li>
<li><p>the number of genes detected</p></li>
<li><p>the percentage of reads mapping to ribosomal</p></li>
<li><p>mitochondrial genes</p></li>
<li><p>ercc recovery (if available)</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>rRNA_genes</strong> (<cite>list</cite> or <cite>str</cite>) – Either a list of rRNA genes or a string containing the path to a file containing
the rRNA genes (one gene per line).</p></li>
<li><p><strong>sd_thres</strong> (<cite>float</cite>) – Number of standard deviations to consider significant, i.e. cells are low quality
if this. Set to higher to remove fewer cells. Default: 3</p></li>
<li><p><strong>seed</strong> (<cite>float</cite>) – For the random number generator. Default: 42</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of low quality cells that were identified, and also modifies the passed
object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.find_mitochondrial_genes">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_mitochondrial_genes</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">mito_pattern='^mt-'</em>, <em class="sig-param">genes=None</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_mitochondrial_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Find mitochondrial genes and adds percent mitochondrial expression of total
expression to the cellular meta data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>mito_pattern</strong> (<cite>str</cite>) – A regular expression matching mitochondrial gene symbols. Default: “^mt-“</p></li>
<li><p><strong>genes</strong> (<cite>list</cite>, optional) – Instead of using <cite>mito_pattern</cite>, specify a <cite>list</cite> of genes that are mitochondrial.</p></li>
<li><p><strong>verbose</strong> (<em>boolean</em>) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of mitochondrial genes detected.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.impute">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">impute</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">filtered=True</em>, <em class="sig-param">res=0.5</em>, <em class="sig-param">drop_thre=0.5</em>, <em class="sig-param">nworkers='auto'</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.impute" title="Permalink to this definition">¶</a></dt>
<dd><p>Impute dropouts using the method described in Li (2018) Nature Communications</p>
<p class="rubric">Notes</p>
<p>Dropouts are artifacts in scRNA-seq data. One method to alleviate the problem with
dropouts is to perform imputation (i.e. replacing missing data points with predicted
values).</p>
<p>The present method uses a different procedure for subpopulation identification
as compared with the original paper.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>filtered</strong> (<cite>bool</cite>) – If data have been filtered using <a class="reference internal" href="#adobo.preproc.simple_filter" title="adobo.preproc.simple_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.preproc.simple_filter()</span></code></a>, run
imputation on filtered data; otherwise runs on the entire raw read count matrix.
Default: True</p></li>
<li><p><strong>res</strong> (<cite>float</cite>) – Resolution parameter for the Leiden clustering, change to modify cluster
resolution. Default: 0.5</p></li>
<li><p><strong>drop_thre</strong> (<cite>float</cite>) – Drop threshold. Default: 0.5</p></li>
<li><p><strong>nworkers</strong> (<cite>int</cite> or <cite>{‘auto’}</cite>) – If a string, then the only accepted value is ‘auto’, and the number of worker
processes will be the total number of detected physical cores. If an integer
then it specifies the number of worker processes. Default: ‘auto’</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>) – Be verbose or not. Default: True</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id50"><span class="brackets">1</span></dt>
<dd><p>Li &amp; Li (2018)
An accurate and robust imputation method scImpute for single-cell
RNA-seq data <a class="reference external" href="https://www.nature.com/articles/s41467-018-03405-7">https://www.nature.com/articles/s41467-018-03405-7</a></p>
</dd>
<dt class="label" id="id51"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://github.com/Vivianstats/scImpute">https://github.com/Vivianstats/scImpute</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.reset_filters">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">reset_filters</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.reset_filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets cell and gene filters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing. Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.simple_filter">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">simple_filter</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">minreads=1000</em>, <em class="sig-param">maxreads=None</em>, <em class="sig-param">minexpgenes=0.001</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.simple_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cells with too few reads and genes with very low expression</p>
<p class="rubric">Notes</p>
<p>The cells are filtered first and then genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>minreads</strong> (<cite>int</cite>, optional) – Minimum number of reads per cell required to keep the cell. Default: 1000</p></li>
<li><p><strong>maxreads</strong> (<cite>int</cite>, optional) – Set a maximum number of reads allowed. Useful for filtering out suspected doublets.</p></li>
<li><p><strong>minexpgenes</strong> (<cite>str</cite>, optional) – If this value is a float, then at least that fraction of cells must express the
gene. If integer, then it denotes the minimum that number of cells must express
the gene. Set to None to ignore this option. Default: 0.001</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">simple_filter</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">minreads</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">maxreads</span><span class="o">=</span><span class="mi">15000</span><span class="p">,</span> <span class="n">minexpgenes</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>int</em> – Number of cells removed.</p></li>
<li><p><em>int</em> – Number of genes removed.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.symbol_switch">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">symbol_switch</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">species</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.symbol_switch" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes gene symbol format</p>
<p class="rubric">Notes</p>
<p>If gene symbols are in the format ENS[0-9]+, this function changes gene identifiers
to symbol_ENS[0-9]+.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>species</strong> (<em>'{'human'</em><em>, </em><em>'mouse'}'</em>) – Species. Default: ‘human’</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exp</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;pbmc8k.mat.gz&#39;</span><span class="p">,</span> <span class="n">bundled</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">symbol_switch</span><span class="p">(</span><span class="n">exp</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.traj">
<span id="adobo-traj-module"></span><h2>adobo.traj module<a class="headerlink" href="#module-adobo.traj" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id52">
<h3>Summary<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h3>
<p>Functions for trajectory analysis.</p>
<dl class="function">
<dt id="adobo.traj.slingshot">
<code class="sig-prename descclassname">adobo.traj.</code><code class="sig-name descname">slingshot</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">name=()</em>, <em class="sig-param">min_cluster_size=10</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.traj.slingshot" title="Permalink to this definition">¶</a></dt>
<dd><p>Trajectory analysis on the cluster level following the strategy in the R
package slingshot</p>
<p class="rubric">Notes</p>
<p>Slingshot’s approach takes cells in a low dimensional space (UMAP is used
below) and a clustering to generate a graph where vertices are clusters.</p>
<p>Only slingthot’s ‘getLineages’ method is used at the moment.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id53"><span class="brackets">1</span></dt>
<dd><p>Street et al. (2018) BMC Genomics. Slingshot: cell lineage and pseudotime
inference for single-cell transcriptomics</p>
</dd>
<dt class="label" id="id54"><span class="brackets">2</span></dt>
<dd><p><a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/slingshot.html">https://bioconductor.org/packages/release/bioc/html/slingshot.html</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>name</strong> (<cite>tuple</cite>) – A tuple of normalization to use. If it has the length zero, then all
available normalizations will be used.</p></li>
<li><p><strong>min_cluster_size</strong> (<cite>int</cite>) – Minimum number of cells per cluster to include the cluster. Default: 10</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not. Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Nothing modifies the passed object.</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-adobo" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="adobo.excepthook">
<code class="sig-prename descclassname">adobo.</code><code class="sig-name descname">excepthook</code><span class="sig-paren">(</span><em class="sig-param">type</em>, <em class="sig-param">value</em>, <em class="sig-param">tb</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.excepthook" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Oscar Franzen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.1.
    </div>
  </body>
</html>