
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>adobo package &#8212; adobo  documentation</title>
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">adobo package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-adobo.IO">adobo.IO module</a><ul>
<li><a class="reference internal" href="#summary">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.data">adobo.data module</a><ul>
<li><a class="reference internal" href="#id1">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.normalize">adobo.normalize module</a><ul>
<li><a class="reference internal" href="#id2">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.plotting">adobo.plotting module</a><ul>
<li><a class="reference internal" href="#id3">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo.preproc">adobo.preproc module</a><ul>
<li><a class="reference internal" href="#id4">Summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-adobo">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/adobo.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="adobo-package">
<h1>adobo package<a class="headerlink" href="#adobo-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-adobo.IO">
<span id="adobo-io-module"></span><h2>adobo.IO module<a class="headerlink" href="#module-adobo.IO" title="Permalink to this headline">¶</a></h2>
<div class="section" id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<p>Functions for reading and writing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.IO.load_from_file">
<code class="sig-prename descclassname">adobo.IO.</code><code class="sig-name descname">load_from_file</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">sep='\t'</em>, <em class="sig-param">header=0</em>, <em class="sig-param">column_id=True</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.IO.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a gene expression matrix consisting of raw read counts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<cite>str</cite>) – Path to the file containing input data. Should be a matrix where
columns are cells and rows are genes.</p></li>
<li><p><strong>sep</strong> (<cite>str</cite>, optional) – Character used to separate fields (default: “\t”).</p></li>
<li><p><strong>header</strong> (<cite>str</cite>, optional) – If the data file has a header. 0 means yes otherwise None (default: <cite>0</cite>).</p></li>
<li><p><strong>column_id</strong> (<cite>bool</cite>, optional) – Whether the header (first line) of the file contains a column ID for the genes. If
this is the case, set this to True, otherwise False (default: True).</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – To be verbose or not (default: False).</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Gene expression matrix should not be normalized.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dataset class object.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.data">
<span id="adobo-data-module"></span><h2>adobo.data module<a class="headerlink" href="#module-adobo.data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>Summary<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>This module contains a data storage class.</p>
<dl class="class">
<dt id="adobo.data.dataset">
<em class="property">class </em><code class="sig-prename descclassname">adobo.data.</code><code class="sig-name descname">dataset</code><span class="sig-paren">(</span><em class="sig-param">raw_mat</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.data.dataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Storage container for raw data and analysis results.</p>
<dl class="attribute">
<dt id="adobo.data.dataset.exp_mat">
<code class="sig-name descname">exp_mat</code><a class="headerlink" href="#adobo.data.dataset.exp_mat" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw read count matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.exp_mito">
<code class="sig-name descname">exp_mito</code><a class="headerlink" href="#adobo.data.dataset.exp_mito" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw read count matrix containing mitochondrial genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.exp_ERCC">
<code class="sig-name descname">exp_ERCC</code><a class="headerlink" href="#adobo.data.dataset.exp_ERCC" title="Permalink to this definition">¶</a></dt>
<dd><p>Raw read count matrix containing ERCC spikes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.low_quality_cells">
<code class="sig-name descname">low_quality_cells</code><a class="headerlink" href="#adobo.data.dataset.low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Low quality cells identified with <a class="reference internal" href="#adobo.preproc.find_low_quality_cells" title="adobo.preproc.find_low_quality_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adobo.preproc.find_low_quality_cells()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.norm">
<code class="sig-name descname">norm</code><a class="headerlink" href="#adobo.data.dataset.norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalized gene expression data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="adobo.data.dataset.norm_method">
<code class="sig-name descname">norm_method</code><a class="headerlink" href="#adobo.data.dataset.norm_method" title="Permalink to this definition">¶</a></dt>
<dd><p>Method used for normalization.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><cite>str</cite></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.normalize">
<span id="adobo-normalize-module"></span><h2>adobo.normalize module<a class="headerlink" href="#module-adobo.normalize" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id2">
<h3>Summary<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>This module contains functions to normalize raw read counts.</p>
<dl class="function">
<dt id="adobo.normalize.full_quantile_normalization">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">full_quantile_normalization</code><span class="sig-paren">(</span><em class="sig-param">d</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.full_quantile_normalization" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs full quantile normalization (FQN).</p>
<p class="rubric">Notes</p>
<p>FQN was a popular normalization scheme for microarray data.
The present function does not handle ties well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>obj</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame object containing raw read counts (rows=genes,
columns=cells).</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>[0] Bolstad et al. (2003) <a class="reference external" href="https://academic.oup.com/bioinformatics/article/19/2/185/372664">https://academic.oup.com/bioinformatics/article/19/2/185/372664</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A normalized data matrix with same dimensions as before.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.normalize.norm">
<code class="sig-prename descclassname">adobo.normalize.</code><code class="sig-name descname">norm</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">method='depth'</em>, <em class="sig-param">log2=True</em>, <em class="sig-param">small_const=1</em>, <em class="sig-param">remove_low_qual_cells=True</em>, <em class="sig-param">exon_lengths=None</em>, <em class="sig-param">scaling_factor=10000</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.normalize.norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalizes gene expression data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A dataset class object.</p></li>
<li><p><strong>method</strong> (<em>{'standard'</em><em>, </em><em>'rpkm'}</em>) – Specifies the method to use. <cite>standard</cite> refers to the simplest normalization
strategy involving scaling genes by total number of reads per cell. <cite>rpkm</cite>
performs RPKM normalization and requires the <cite>exon_lengths</cite> parameter to be set.</p></li>
<li><p><strong>log2</strong> (<cite>bool</cite>) – Perform log2 transformation (default: True)</p></li>
<li><p><strong>small_const</strong> (<cite>float</cite>) – A small constant to add to expression values to avoid log’ing genes with zero
expression (default: 1).</p></li>
<li><p><strong>remove_low_qual_cells</strong> (<cite>bool</cite>) – Remove low quality cells identified using <a class="reference internal" href="#adobo.preproc.find_low_quality_cells" title="adobo.preproc.find_low_quality_cells"><code class="xref py py-meth docutils literal notranslate"><span class="pre">adobo.preproc.find_low_quality_cells()</span></code></a>.</p></li>
<li><p><strong>exon_lengths</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>) – A pandas data frame containing two columns; first column should be gene names
matching the data matrix and the second column should contain exon lengths.</p></li>
<li><p><strong>scaling_factor</strong> (<cite>int</cite>) – Scaling factor used to multiply the scaled counts with. Only used for
<cite>method=”depth”</cite> (default: 10000).</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>[0] Cole et al. (2019) <a class="reference external" href="https://www.biorxiv.org/content/10.1101/235382v2">https://www.biorxiv.org/content/10.1101/235382v2</a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.plotting">
<span id="adobo-plotting-module"></span><h2>adobo.plotting module<a class="headerlink" href="#module-adobo.plotting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id3">
<h3>Summary<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>Functions for plotting scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.plotting.barplot_genes_per_cell">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">barplot_genes_per_cell</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">barcolor='#E69F00'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">title='expressed genes'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.barplot_genes_per_cell" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a bar plot of number of expressed genes per cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object</p></li>
<li><p><strong>barcolor</strong> (<cite>str</cite>, optional) – Color of the bars (default: “#E69F00”).</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file.</p></li>
<li><p><strong>title</strong> (<cite>str</cite>, optional) – Title of the plot (default: “sequencing reads”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.plotting.barplot_reads_per_cell">
<code class="sig-prename descclassname">adobo.plotting.</code><code class="sig-name descname">barplot_reads_per_cell</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">barcolor='#E69F00'</em>, <em class="sig-param">filename=None</em>, <em class="sig-param">title='sequencing reads'</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.plotting.barplot_reads_per_cell" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a bar plot of read counts per cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>barcolor</strong> (<cite>str</cite>, optional) – Color of the bars (default: “#E69F00”).</p></li>
<li><p><strong>filename</strong> (<cite>str</cite>, optional) – Write plot to file.</p></li>
<li><p><strong>title</strong> (<cite>str</cite>, optional) – Title of the plot (default: “sequencing reads”).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo.preproc">
<span id="adobo-preproc-module"></span><h2>adobo.preproc module<a class="headerlink" href="#module-adobo.preproc" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id4">
<h3>Summary<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>Functions for pre-processing scRNA-seq data.</p>
<dl class="function">
<dt id="adobo.preproc.detect_ERCC_spikes">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">detect_ERCC_spikes</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">ERCC_pattern='^ERCC[_-]\\S+$'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.detect_ERCC_spikes" title="Permalink to this definition">¶</a></dt>
<dd><p>Moves ERCC (if present) to a separate container</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>ERCC_pattern</strong> (<cite>str</cite>, optional) – A regular expression matching ERCC gene symbols (default: “ERCC[_-]S+$”).</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of detected ERCC spikes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.detect_mito">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">detect_mito</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">mito_pattern='^mt-'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.detect_mito" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove mitochondrial genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>mito_pattern</strong> (<cite>str</cite>, optional) – A regular expression matching mitochondrial gene symbols (default: “^mt-“)</p></li>
<li><p><strong>verbose</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Be verbose or not (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Number of mitochondrial genes detected.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.find_low_quality_cells">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">find_low_quality_cells</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">rRNA_genes</em>, <em class="sig-param">sd_thres=3</em>, <em class="sig-param">seed=42</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.find_low_quality_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Statistical detection of low quality cells</p>
<p class="rubric">Notes</p>
<p>Mahalanobis distances are computed from five quality metrics. A robust estimate of
covariance is used in the Mahalanobis function. Cells with Mahalanobis distances of
three standard deviations from the mean are by default considered outliers.
The five metrics are:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>log-transformed number of molecules detected</p></li>
<li><p>the number of genes detected</p></li>
<li><p>the percentage of reads mapping to ribosomal</p></li>
<li><p>mitochondrial genes</p></li>
<li><p>ERCC recovery (if available)</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>rRNA_genes</strong> (<cite>list</cite>) – List of rRNA genes.</p></li>
<li><p><strong>sd_thres</strong> (<cite>float</cite>, optional) – Number of standard deviations to consider significant, i.e. cells are low quality
if this. Set to higher to remove fewer cells (default: 3).</p></li>
<li><p><strong>seed</strong> (<cite>float</cite>, optional) – For the random number generator (default: 42).</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of low quality cells that were identified.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.remove_empty">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">remove_empty</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.remove_empty" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes empty cells and genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>verbose</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Be verbose or not (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>int</em> – Number of empty cells removed.</p></li>
<li><p><em>int</em> – Number of empty genes removed.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="adobo.preproc.simple_filter">
<code class="sig-prename descclassname">adobo.preproc.</code><code class="sig-name descname">simple_filter</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">minreads=1000</em>, <em class="sig-param">minexpgenes=0.001</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#adobo.preproc.simple_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes cells with too few reads and genes with very low expression</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> (<a class="reference internal" href="#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>) – A data class object.</p></li>
<li><p><strong>minreads</strong> (<cite>int</cite>, optional) – Minimum number of reads per cell required to keep the cell (default: 1000).</p></li>
<li><p><strong>minexpgenes</strong> (<cite>str</cite>, optional) – If this value is a float, then at least that fraction of cells must express the
gene. If integer, then it denotes the minimum that number of cells must express
the gene (default: 0.001).</p></li>
<li><p><strong>verbose</strong> (<cite>bool</cite>, optional) – Be verbose or not (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>int</em> – Number of cells removed.</p></li>
<li><p><em>int</em> – Number of genes removed.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="module-adobo">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-adobo" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Oscar Franzen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>