
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Welcome to adobo’s documentation! &#8212; adobo  documentation</title>
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to adobo’s documentation!</a><ul>
<li><a class="reference internal" href="#what-adobo-is">What adobo is</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#getting-started">Getting started</a><ul>
<li><a class="reference internal" href="#loading-the-package">Loading the package</a></li>
<li><a class="reference internal" href="#loading-your-data-from-a-text-file">Loading your data from a text file</a></li>
<li><a class="reference internal" href="#getting-detailed-help">Getting detailed help</a></li>
<li><a class="reference internal" href="#loading-compressed-data">Loading compressed data</a></li>
<li><a class="reference internal" href="#data-examination">Data examination</a></li>
<li><a class="reference internal" href="#cleaning-your-data">Cleaning your data</a></li>
<li><a class="reference internal" href="#detecting-ercc-spikes">Detecting ERCC spikes</a></li>
<li><a class="reference internal" href="#detecting-mitochondrial-genes">Detecting mitochondrial genes</a></li>
<li><a class="reference internal" href="#applying-simple-filters">Applying simple filters</a></li>
<li><a class="reference internal" href="#automatic-detection-of-low-quality-cells">Automatic detection of low quality cells</a></li>
</ul>
</li>
<li><a class="reference internal" href="#normalization">Normalization</a></li>
<li><a class="reference internal" href="#examining-analysis-history">Examining analysis history</a></li>
<li><a class="reference internal" href="#detection-of-highly-variable-genes">Detection of highly variable genes</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-adobo-s-documentation">
<h1>Welcome to adobo’s documentation!<a class="headerlink" href="#welcome-to-adobo-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="what-adobo-is">
<h2>What adobo is<a class="headerlink" href="#what-adobo-is" title="Permalink to this headline">¶</a></h2>
<p>adobo is an analysis framework consisting of a set of Python modules to enable exploratory analysis of single cell RNA-seq data (<a class="reference external" href="https://en.wikipedia.org/wiki/Single_cell_sequencing">scRNA-seq</a>). adobo can be used to compose scripts, used in interactive workflows and much more. adobo is developed by Oscar Franzén (limited support is available over e-mail: <a class="reference external" href="mailto:p&#46;oscar&#46;franzen&#37;&#52;&#48;gmail&#46;com">p<span>&#46;</span>oscar<span>&#46;</span>franzen<span>&#64;</span>gmail<span>&#46;</span>com</a>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>adobo is developed and tested on Python version 3.5.3 - older versions may work but are untested.</p>
</div>
<div class="section" id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h3>
<p>The easiest way to install adobo is to clone the GitHub repository:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ git clone https://github.com/oscar-franzen/adobo.git
</pre></div>
</div>
<p>Support for <code class="docutils literal notranslate"><span class="pre">pip3</span></code> installation is on the TODO list.</p>
</div>
<div class="section" id="getting-started">
<h3>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h3>
<div class="section" id="loading-the-package">
<h4>Loading the package<a class="headerlink" href="#loading-the-package" title="Permalink to this headline">¶</a></h4>
<p>The first step is to load the adobo package by importing it:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">adobo</span> <span class="k">as</span> <span class="nn">ad</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Debug information in the form of traceback output is suppressed by default. However, this information is often useful when trying to solve program bugs. To enable full traceback set:</p>
<p><code class="docutils literal notranslate"><span class="pre">ad.debug=1</span></code></p>
</div>
</div>
<div class="section" id="loading-your-data-from-a-text-file">
<h4>Loading your data from a text file<a class="headerlink" href="#loading-your-data-from-a-text-file" title="Permalink to this headline">¶</a></h4>
<p>First we need to create a new adobo object, which is an instance of the class <a class="reference internal" href="adobo.html#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>. This will be a new object containing your single cell data. The input file should be a gene expression matrix (rows as genes and cells as columns) in plain text format. Fields can be separated by any character (default is tab) and it can be changed with the <code class="docutils literal notranslate"><span class="pre">sep</span></code> argument. The data matrix file can have a header or not (<code class="docutils literal notranslate"><span class="pre">header=0</span></code> indicates a header is present, otherwise use <code class="docutils literal notranslate"><span class="pre">header=None</span></code>). If the header of your gene expression matrix contains a label for gene symbols, set <code class="docutils literal notranslate"><span class="pre">column_id='yes'</span></code> or simply <code class="docutils literal notranslate"><span class="pre">column_id='auto'</span></code> to autodetect this. <a class="reference internal" href="adobo.html#adobo.IO.load_from_file" title="adobo.IO.load_from_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.IO.load_from_file()</span></code></a> calls <code class="xref py py-func docutils literal notranslate"><span class="pre">pandas.DataFrame.read_csv()</span></code> and any additional arguments are passed into this function. The function <a class="reference internal" href="adobo.html#adobo.IO.load_from_file" title="adobo.IO.load_from_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.IO.load_from_file()</span></code></a> is used to load data from a raw read count matrix and the returned object is an instance of <a class="reference internal" href="adobo.html#adobo.data.dataset" title="adobo.data.dataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">adobo.data.dataset</span></code></a>:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;GSE95315.tab&#39;</span><span class="p">,</span> <span class="n">column_id</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The loaded data must <strong>not</strong> be normalized, i.e. it should be raw read counts.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All downstream operations and analyses are performed and stored in the adobo object, i.e. functions are applied on this object.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Most adobo functions also have a <code class="docutils literal notranslate"><span class="pre">verbose</span></code> argument, which when <code class="docutils literal notranslate"><span class="pre">True</span></code> makes the function more informative.</p>
</div>
</div>
<div class="section" id="getting-detailed-help">
<h4>Getting detailed help<a class="headerlink" href="#getting-detailed-help" title="Permalink to this headline">¶</a></h4>
<p>All functions in adobo have extensive documentation, which are accessible as docstrings on the Python interactive console as well as online:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">ad</span><span class="p">)</span>
<span class="n">help</span><span class="p">(</span><span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="loading-compressed-data">
<h4>Loading compressed data<a class="headerlink" href="#loading-compressed-data" title="Permalink to this headline">¶</a></h4>
<p>scRNA-seq is zero inflated, and therefore compress strongly. We can load the compressed data directly without having to uncompress it first; the compression format is detected automatically (gzip, bz2, zip and xz are supported). For example:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">IO</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s1">&#39;GSE95315.tab.gz&#39;</span><span class="p">,</span> <span class="n">column_id</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Your gene expression data is stored in the attribute <code class="docutils literal notranslate"><span class="pre">data.exp_mat</span></code>, and after loading it is good to examine that the data were loaded properly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span>
<span class="go">53,889 genes and 384 cells were loaded</span>
</pre></div>
</div>
</div>
<div class="section" id="data-examination">
<h4>Data examination<a class="headerlink" href="#data-examination" title="Permalink to this headline">¶</a></h4>
<p>It’s can be useful to examine number of reads per cell in a barplot format:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">barplot_reads_per_cell</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<p>Which will generate the plot:</p>
<img alt="_images/2019-08-30-131424_587x441_scrot.png" src="_images/2019-08-30-131424_587x441_scrot.png" />
<p>It is also a good idea to examine number of expressed genes per cell:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">barplot_genes_per_cell</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/2019-08-30-131930_594x451_scrot.png" src="_images/2019-08-30-131930_594x451_scrot.png" />
</div>
<div class="section" id="cleaning-your-data">
<h4>Cleaning your data<a class="headerlink" href="#cleaning-your-data" title="Permalink to this headline">¶</a></h4>
<p>Before we proceed to normalization we should start with some data cleaning. A good first step is to look for empty cells and genes. Empty genes/cells do obviously not carry any information and can be removed:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">remove_empty</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="detecting-ercc-spikes">
<h4>Detecting ERCC spikes<a class="headerlink" href="#detecting-ercc-spikes" title="Permalink to this headline">¶</a></h4>
<p>ERCC are known amounts of synthetic constructs added to RNA-seq libraries for quality control and normalization purposes <a class="reference internal" href="#jiang2011" id="id1">[1]</a>. Not all experiments use ERCC spikes, but many do. The ERCC “genes” are usually prefixed with <cite>ERCC-</cite> in the gene expression matrix. This function is used to put them aside so that they are not included in downstream analyses.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ERCC_pattern</span></code> argument can be used to set a regular expression for detecting these “genes”.</p>
</div>
<p>The <a class="reference internal" href="adobo.html#adobo.preproc.detect_ERCC_spikes" title="adobo.preproc.detect_ERCC_spikes"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.preproc.detect_ERCC_spikes()</span></code></a> is used to filter out the ERCC (stored in <a class="reference internal" href="adobo.html#adobo.data.dataset._exp_ERCC" title="adobo.data.dataset._exp_ERCC"><code class="xref py py-attr docutils literal notranslate"><span class="pre">adobo.data.dataset._exp_ERCC</span></code></a>):</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">detect_ERCC_spikes</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="detecting-mitochondrial-genes">
<h4>Detecting mitochondrial genes<a class="headerlink" href="#detecting-mitochondrial-genes" title="Permalink to this headline">¶</a></h4>
<p>It’s often a good idea not to include the mitochondrial genes in the analysis and some downstream analyses require that adobo knows which genes are mitochondrial. Usually the mitochondrial genes in human and mouse genomes have gene symbols starting with the prefix <code class="docutils literal notranslate"><span class="pre">mt-</span></code>, but this might vary from species to species.</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">detect_mito</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<p>Or specify a custom regular expression to identify them:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">detect_mito</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">mito_pattern</span><span class="o">=</span><span class="s1">&#39;^mito-&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="applying-simple-filters">
<h4>Applying simple filters<a class="headerlink" href="#applying-simple-filters" title="Permalink to this headline">¶</a></h4>
<p>Simple filters refers to applying a strict minimum cutoff on the number of expressed genes per cell and the total read depth per cell. Simple filters are usually effective in removing low quality cells and uninformative genes. If your data come from Drop-seq, 10X, etc, requiring at least 1000 uniquely mapped reads per cell is often sufficient:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">simple_filter</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">minreads</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">minexpgenes</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If your protocol is applying full-length mRNA sequencing, e.g. SMART-seq2, then your <code class="docutils literal notranslate"><span class="pre">minreads</span></code> threshold should be higher, for example 10000.</p>
</div>
<p>It is also desirable to remove genes with an expression signal in very few cells; such genes may contribute more noise than information. The <code class="docutils literal notranslate"><span class="pre">minexpgenes</span></code> argument can be used to control how genes are filtered out. If you wish to not remove any genes at all, simply set it to zero:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">simple_filter</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">minreads</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">minexpgenes</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Setting <code class="docutils literal notranslate"><span class="pre">minexpgenes</span></code> to a fraction indicates that at least that fraction of cells must express any gene. If <code class="docutils literal notranslate"><span class="pre">minexpgenes</span></code> is an integer it refers to the absolute number of cells that at minimum must express the gene for the gene not to be filtered out.</p>
</div>
<div class="section" id="automatic-detection-of-low-quality-cells">
<h4>Automatic detection of low quality cells<a class="headerlink" href="#automatic-detection-of-low-quality-cells" title="Permalink to this headline">¶</a></h4>
<p>A more sophisticated approach to detection of low quality cells is to use the function <a class="reference internal" href="adobo.html#adobo.preproc.find_low_quality_cells" title="adobo.preproc.find_low_quality_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.preproc.find_low_quality_cells()</span></code></a>, which uses <a class="reference external" href="https://en.wikipedia.org/wiki/Mahalanobis_distance">Mahalanobis distance</a> to identify bad cells from five quality metrics.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p><code class="docutils literal notranslate"><span class="pre">find_low_quality_cells</span></code> requires that there are ERCC spikes in your data.</p>
</div>
<p>The argument <code class="docutils literal notranslate"><span class="pre">rRNA_genes</span></code> should either be a string containing the full path to a file on disk contaiing genes that are rRNA genes (the file should have one gene per line). <code class="docutils literal notranslate"><span class="pre">rRNA_genes</span></code> can also be a <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code> object with gene symbols.</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">find_low_quality_cells</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">rRNA_genes</span><span class="o">=</span><span class="n">rRNA</span><span class="p">)</span>
</pre></div>
</div>
<p>Like all adobo functions, <code class="docutils literal notranslate"><span class="pre">find_low_quality_cells</span></code> modifies the passed object. However, <code class="docutils literal notranslate"><span class="pre">find_low_quality_cells</span></code> also returns a list of cells that are classified as low quality; to prevent such behavior simply assign the return to a variable:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">low_q_cells</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">preproc</span><span class="o">.</span><span class="n">find_low_quality_cells</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">rRNA_genes</span><span class="o">=</span><span class="n">rRNA</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="normalization">
<h3>Normalization<a class="headerlink" href="#normalization" title="Permalink to this headline">¶</a></h3>
<p>Normalization removes technical and sometimes experimental biases and is always necessary prior to analysis. Because a single normalization scheme for scRNA-seq data is not available, adobo supports several different procedures:</p>
<dl class="simple">
<dt><strong>standard</strong></dt><dd><p>Performs a standard normalization by scaling with the total read depth per cell and then multiplying with a scaling factor.</p>
</dd>
<dt><strong>rpkm</strong></dt><dd><p>Normalizes read counts as Reads per kilo base per million mapped reads (RPKM) <a class="reference internal" href="#conesa2016" id="id2">[2]</a>. This method should be used if you need to adjust for gene length, such as in a SMART-Seq2 protocol. To use this procedure you must first prepare a file containing combined exon lengths for genes; the file should contain two columns, <strong>without a header</strong>, and columns separated by one space. The following columns must be present: (1) gene symbols and (2) the sum of exon lengths.</p>
</dd>
<dt><strong>fqn</strong></dt><dd><p>Performs full quantile normalization <a class="reference internal" href="#bolstad2003" id="id3">[3]</a>. FQN was a popular normalization scheme for microarray data. It is not very common in single cell analysis despite having been shown to perform well <a class="reference internal" href="#cole2018" id="id4">[4]</a>. The present implementation does not handle ties well.</p>
</dd>
<dt><strong>clr</strong></dt><dd><p>Centered log ratio normalization. This normalization scheme was introduced in Seurat version 3.0 <a class="reference internal" href="#stuart2018" id="id5">[5]</a>. It is a simple normalization scheme and is an alternative to <code class="docutils literal notranslate"><span class="pre">standard</span></code>.</p>
</dd>
<dt><strong>vsn</strong></dt><dd><p>Variance stabilizing normaliztion based on a negative binomial regression model with regularized parameters. Introduced by <a class="reference internal" href="#hafemeister2019" id="id6">[6]</a> and represents the most complex scheme of the above; appears to marginally improve resolution.</p>
</dd>
</dl>
<p>All normalization schemes, except <code class="docutils literal notranslate"><span class="pre">vsn</span></code>, can be followed by log (base 2) transformation (set by the <code class="docutils literal notranslate"><span class="pre">log2=True</span></code> flag, which is enabled by default).</p>
<p>To perform a <code class="docutils literal notranslate"><span class="pre">standard</span></code> normalization followed by <code class="docutils literal notranslate"><span class="pre">log2</span></code> transformation, run:</p>
<div class="highlight-python3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad</span><span class="o">.</span><span class="n">normalize</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;standard&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you have previously executed <a class="reference internal" href="adobo.html#adobo.preproc.detect_ERCC_spikes" title="adobo.preproc.detect_ERCC_spikes"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.preproc.detect_ERCC_spikes()</span></code></a>, ERCC spikes will be normalized too, and these can be found in <code class="xref py py-attr docutils literal notranslate"><span class="pre">adobo.data.dataset.norm_ERCC</span></code>.</p>
</div>
</div>
<div class="section" id="examining-analysis-history">
<h3>Examining analysis history<a class="headerlink" href="#examining-analysis-history" title="Permalink to this headline">¶</a></h3>
<p>Downstream analyses are performed on the data object. At any time it’s possible to examine what functions have been applied on data object by calling <a class="reference internal" href="adobo.html#adobo.data.dataset.assays" title="adobo.data.dataset.assays"><code class="xref py py-func docutils literal notranslate"><span class="pre">adobo.data.dataset.assays()</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">assays</span><span class="p">()</span>
<span class="go">Number of mitochondrial genes found: 0</span>
<span class="go">Number of ERCC spikes found: 92</span>
<span class="go">Normalization method: &lt;not performed yet&gt;</span>
<span class="go">Has HVG discovery been performed? No</span>
</pre></div>
</div>
</div>
<div class="section" id="detection-of-highly-variable-genes">
<h3>Detection of highly variable genes<a class="headerlink" href="#detection-of-highly-variable-genes" title="Permalink to this headline">¶</a></h3>
<p>Most algorithms used in scRNA-seq analysis performs better when performed only on a subset of measured genes <a class="reference internal" href="#yip2018" id="id7">[7]</a>; this subset is for the most time selected as highly variable genes (HVG). adobo currently implements the following strategies for HVG discovery:</p>
<dl class="simple">
<dt><strong>seurat</strong></dt><dd><p>The function bins the genes according to average expression, then calculates dispersion for each bin as variance to mean ratio. Within each bin, Z-scores are calculated and returned. Z-scores are ranked and the top 1000 are selected. Input data should be normalized first. This strategy was introduced in Seruat <a class="reference internal" href="#stuart2018" id="id8">[5]</a>, it is simple yet highly effective in identifying HVG.</p>
</dd>
<dt><strong>brennecke</strong></dt><dd><p>Implements the method described in <a class="reference internal" href="#brennecke2013" id="id9">[8]</a>. <code class="docutils literal notranslate"><span class="pre">brennecke</span></code> estimates and fits technical noise using ERCC spikes (technical genes) by fitting a generalized linear model with a gamma function and identity link and the parameterization w=a_1+u+a0. It then uses the chi2 distribution to test the null hypothesis that the squared coefficient of variation does not exceed a certain minimum. False discovery rate (FDR)&lt;0.10 is considered significant.</p>
</dd>
</dl>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h3>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
</ul>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<p id="bibtex-bibliography-index-0"><dl class="citation">
<dt class="label" id="jiang2011"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Lichun Jiang, Felix Schlesinger, Carrie A. Davis, Yu Zhang, Renhua Li, Marc Salit, Thomas R. Gingeras, and Brian Oliver. Synthetic spike-in standards for RNA-seq experiments. <em>Genome Research</em>, 21(9):1543–1551, September 2011. <a class="reference external" href="https://doi.org/10.1101/gr.121095.111">doi:10.1101/gr.121095.111</a>.</p>
</dd>
<dt class="label" id="conesa2016"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p>Ana Conesa, Pedro Madrigal, Sonia Tarazona, David Gomez-Cabrero, Alejandra Cervera, Andrew McPherson, Michał Wojciech Szcześniak, Daniel J. Gaffney, Laura L. Elo, Xuegong Zhang, and Ali Mortazavi. A survey of best practices for RNA-seq data analysis. <em>Genome Biology</em>, 17(1):13, January 2016. <a class="reference external" href="https://doi.org/10.1186/s13059-016-0881-8">doi:10.1186/s13059-016-0881-8</a>.</p>
</dd>
<dt class="label" id="bolstad2003"><span class="brackets"><a class="fn-backref" href="#id3">3</a></span></dt>
<dd><p>B.M. Bolstad, R.A Irizarry, M. Åstrand, and T.P. Speed. A comparison of normalization methods for high density oligonucleotide array data based on variance and bias. <em>Bioinformatics</em>, 19(2):185–193, 01 2003. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/19.2.185">doi:10.1093/bioinformatics/19.2.185</a>.</p>
</dd>
<dt class="label" id="cole2018"><span class="brackets"><a class="fn-backref" href="#id4">4</a></span></dt>
<dd><p>Michael B. Cole, Davide Risso, Allon Wagner, David DeTomaso, John Ngai, Elizabeth Purdom, Sandrine Dudoit, and Nir Yosef. Performance assessment and selection of normalization procedures for single-cell rna-seq. <em>bioRxiv</em>, 2018. <a class="reference external" href="https://doi.org/10.1101/235382">doi:10.1101/235382</a>.</p>
</dd>
<dt class="label" id="stuart2018"><span class="brackets">5</span><span class="fn-backref">(<a href="#id5">1</a>,<a href="#id8">2</a>)</span></dt>
<dd><p>Tim Stuart, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Marlon Stoeckius, Peter Smibert, and Rahul Satija. Comprehensive integration of single cell data. <em>bioRxiv</em>, 2018. <a class="reference external" href="https://doi.org/10.1101/460147">doi:10.1101/460147</a>.</p>
</dd>
<dt class="label" id="hafemeister2019"><span class="brackets"><a class="fn-backref" href="#id6">6</a></span></dt>
<dd><p>Christoph Hafemeister and Rahul Satija. Normalization and variance stabilization of single-cell rna-seq data using regularized negative binomial regression. <em>bioRxiv</em>, 2019. <a class="reference external" href="https://doi.org/10.1101/576827">doi:10.1101/576827</a>.</p>
</dd>
<dt class="label" id="yip2018"><span class="brackets"><a class="fn-backref" href="#id7">7</a></span></dt>
<dd><p>Shun H Yip, Pak Chung Sham, and Junwen Wang. Evaluation of tools for highly variable gene discovery from single-cell RNA-seq data. <em>Briefings in Bioinformatics</em>, 02 2018. <a class="reference external" href="https://doi.org/10.1093/bib/bby011">doi:10.1093/bib/bby011</a>.</p>
</dd>
<dt class="label" id="brennecke2013"><span class="brackets"><a class="fn-backref" href="#id9">8</a></span></dt>
<dd><p>Philip Brennecke, Simon Anders, Jong Kyoung Kim, Aleksandra A Kołodziejczyk, Xiuwei Zhang, Valentina Proserpio, Bianka Baying, Vladimir Benes, Sarah A Teichmann, John C Marioni, and Marcus G Heisler. Accounting for technical noise in single-cell RNA-seq experiments. <em>Nature Methods</em>, 10:1093, September 2013. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.2645">https://doi.org/10.1038/nmeth.2645</a>.</p>
</dd>
</dl>
</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">adobo  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Oscar Franzen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>